diff --git a/CMakeLists.txt b/CMakeLists.txt
index f37cdae2a..c548b8e33 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -157,6 +157,16 @@ if(WITH_RDMA)
   set(HAVE_RDMA TRUE)
 endif()
 
+option(WITH_UCX "Enable UCX async messenger" OFF)
+if(WITH_UCX)
+#    message(FATAL_ERROR "With UCX is ${WITH_UCX}")
+  find_package(UCX REQUIRED)
+  set(HAVE_UCX ${UCX_FOUND})
+  message(STATUS "UCX_LIBS: ${UCX_LIBS}")
+else()
+    message(STATUS "UCX support is disabled")
+endif(WITH_UCX)
+
 find_package(Backtrace)
 
 option(WITH_RBD "Enable RADOS Block Device related targets" ON)
diff --git a/ceph.spec b/ceph.spec
index accc10eb9..296eff6df 100644
--- a/ceph.spec
+++ b/ceph.spec
@@ -1411,7 +1411,8 @@ cmake .. \
     -DWITH_SEASTAR:BOOL=ON \
     -DWITH_JAEGER:BOOL=OFF \
 %endif
-    -DWITH_GRAFANA:BOOL=ON
+    -DWITH_GRAFANA:BOOL=ON \
+    -DWITH_UCX=/usr
 
 %if %{with cmake_verbose_logging}
 cat ./CMakeFiles/CMakeOutput.log
diff --git a/ceph.spec.in b/ceph.spec.in
index e4cbbef69..9eb25eb04 100644
--- a/ceph.spec.in
+++ b/ceph.spec.in
@@ -1411,7 +1411,8 @@ cmake .. \
     -DWITH_SEASTAR:BOOL=ON \
     -DWITH_JAEGER:BOOL=OFF \
 %endif
-    -DWITH_GRAFANA:BOOL=ON
+    -DWITH_GRAFANA:BOOL=ON \
+    -DWITH_UCX=/usr
 
 %if %{with cmake_verbose_logging}
 cat ./CMakeFiles/CMakeOutput.log
diff --git a/cmake/modules/FindUCX.cmake b/cmake/modules/FindUCX.cmake
new file mode 100644
index 000000000..e23f71bfb
--- /dev/null
+++ b/cmake/modules/FindUCX.cmake
@@ -0,0 +1,37 @@
+# - Find ucx
+# Find the ucx library and includes
+#
+# UCX_INCLUDE_DIR - where to find ucp.h
+# UCX_LIBS - List of libraries when using ucx 
+# UCX_FOUND - True if ucx found.
+
+# sanity check
+
+message(STATUS "Looking for UCX in ${WITH_UCX}")
+if (WITH_UCX AND EXISTS ${WITH_UCX})
+    find_path(UCX_INCLUDE_DIR ucp/api/ucp.h HINTS "${WITH_UCX}/include")
+    find_library(UCP_LIBRARY ucp HINTS "${WITH_UCX}/lib")
+    find_library(UCS_LIBRARY ucs HINTS "${WITH_UCX}/lib")
+endif ()
+
+message (STATUS "ucx include ${UCX_INCLUDE_DIR}")
+
+message (STATUS "ucp lib ${UCP_LIBRARY}")
+message (STATUS "ucs lib ${UCS_LIBRARY}")
+
+if (UCP_LIBRARY)
+    if (WITH_UCX_DEBUG AND (${WITH_UCX_DEBUG} MATCHES "yes"))
+        set(UCX_DEBUG_FLAGS TRUE)
+    endif ()
+    set(UCX_LIBS
+        ${UCP_LIBRARY}
+        ${UCS_LIBRARY})
+endif (UCP_LIBRARY)
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(ucx DEFAULT_MSG UCX_LIBS UCX_INCLUDE_DIR)
+
+mark_as_advanced(
+    UCX_LIBS
+    UCX_INCLUDE_DIR
+)
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index cdc183d24..748c01da8 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -276,6 +276,16 @@ if(WITH_CEPHFS_JAVA)
   add_subdirectory(java)
 endif()
 
+if(HAVE_UCX)
+    if (UCX_DEBUG_FLAGS)
+          set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I${UCX_INCLUDE_DIR} -g -O0")
+          message(STATUS " adding UCX debugging flags...")
+    else ()
+          set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I${UCX_INCLUDE_DIR}")
+          message(STATUS " adding UCX compilation flags...")
+    endif()
+endif(HAVE_UCX)
+
 if (WITH_BLKIN)
   add_subdirectory(blkin/blkin-lib)
 endif(WITH_BLKIN)
@@ -466,6 +476,10 @@ if(HAVE_QATZIP)
   list(APPEND ceph_common_deps ${qatzip_LIBRARIES})
 endif()
 
+if(HAVE_UCX)
+  list(APPEND ceph_common_deps ${UCX_LIBS})
+endif()
+
 if(WITH_DPDK)
   list(APPEND ceph_common_deps common_async_dpdk)
 endif()
diff --git a/src/ceph-volume/ceph_volume/util/encryption.py b/src/ceph-volume/ceph_volume/util/encryption.py
index 0831d8597..166c2c34c 100644
--- a/src/ceph-volume/ceph_volume/util/encryption.py
+++ b/src/ceph-volume/ceph_volume/util/encryption.py
@@ -152,8 +152,7 @@ def get_dmcrypt_key(osd_id, osd_fsid, lockbox_keyring=None):
     )
     if returncode != 0:
         raise RuntimeError('Unable to retrieve dmcrypt secret')
-    return ' '.join(stdout).strip()
-
+    return ' '.join([i for i in stdout if not i.startswith("#")]).strip()
 
 def write_lockbox_keyring(osd_id, osd_fsid, secret):
     """
diff --git a/src/ceph-volume/ceph_volume/util/prepare.py b/src/ceph-volume/ceph_volume/util/prepare.py
index ff7427eed..f18e3997d 100644
--- a/src/ceph-volume/ceph_volume/util/prepare.py
+++ b/src/ceph-volume/ceph_volume/util/prepare.py
@@ -174,8 +174,7 @@ def create_id(fsid, json_secrets, osd_id=None):
     )
     if returncode != 0:
         raise RuntimeError('Unable to create a new OSD id')
-    return ' '.join(stdout).strip()
-
+    return ' '.join([i for i in stdout if not i.startswith("#")]).strip()
 
 def osd_id_available(osd_id):
     """
diff --git a/src/ceph_mgr.cc b/src/ceph_mgr.cc
index 67bda0c51..244b49f9f 100644
--- a/src/ceph_mgr.cc
+++ b/src/ceph_mgr.cc
@@ -62,6 +62,7 @@ int main(int argc, const char **argv)
   pick_addresses(g_ceph_context, CEPH_PICK_ADDRESS_PUBLIC);
 
   global_init_daemonize(g_ceph_context);
+  g_ceph_context->notify_pre_fork();
   global_init_chdir(g_ceph_context);
   common_init_finish(g_ceph_context);
 
diff --git a/src/cephadm/cephadm b/src/cephadm/cephadm
index eaaeb8406..e59cf80b6 100644
--- a/src/cephadm/cephadm
+++ b/src/cephadm/cephadm
@@ -3860,6 +3860,9 @@ Before=ceph-{fsid}.target
 [Service]
 LimitNOFILE=1048576
 LimitNPROC=1048576
+LimitMEMLOCK=infinity
+LimitCORE=infinity
+PrivateDevices=no
 EnvironmentFile=-/etc/environment
 ExecStart=/bin/bash {data_dir}/{fsid}/%i/unit.run
 ExecStop=-/bin/bash -c 'bash {data_dir}/{fsid}/%i/unit.stop'
@@ -4023,7 +4026,7 @@ class CephContainer:
             cmd_args.append('--net=host')
         if self.entrypoint:
             cmd_args.extend(['--entrypoint', self.entrypoint])
-        if self.privileged:
+        if self.privileged or not self.privileged:
             cmd_args.extend([
                 '--privileged',
                 # let OSD etc read block devs that haven't been chowned
diff --git a/src/common/options/global.yaml.in b/src/common/options/global.yaml.in
index a0ebf3576..855da075b 100644
--- a/src/common/options/global.yaml.in
+++ b/src/common/options/global.yaml.in
@@ -875,7 +875,7 @@ options:
   fmt_desc: Transport type used by Async Messenger. Can be ``async+posix``,
     ``async+dpdk`` or ``async+rdma``. Posix uses standard TCP/IP networking and is
     default. Other transports may be experimental and support may be limited.
-  default: async+posix
+  default: async+ucx
   flags:
   - startup
   with_legacy: true
@@ -883,6 +883,7 @@ options:
   type: str
   level: advanced
   desc: Messenger implementation to use for the public network
+  default: async+ucx
   long_desc: If not specified, use ms_type
   see_also:
   - ms_type
@@ -893,6 +894,7 @@ options:
   type: str
   level: advanced
   desc: Messenger implementation to use for the internal cluster network
+  default: async+ucx
   long_desc: If not specified, use ms_type
   see_also:
   - ms_type
@@ -1415,6 +1417,62 @@ options:
   level: advanced
   default: ib
   with_legacy: true
+- name: ms_async_ucx_device
+  type: str
+  level: advanced
+  default: all
+  with_legacy: true
+- name: ms_async_ucx_tls
+  type: str
+  level: advanced
+  default: tcp,self
+  with_legacy: true
+- name: ms_async_ucx_rndv_thresh
+  type: str
+  level: advanced
+  default: 8k
+  with_legacy: true
+- name: ms_async_ucx_log_level
+  type: str
+  level: advanced
+  default: ERROR
+  with_legacy: true
+- name: rdma_cm_source_addr
+  type: str
+  level: advanced
+  default: 
+  with_legacy: true
+- name: ms_async_ucx_max_recv
+  type: uint
+  level: advanced
+  desc: ucx maximum number of concurrent RDMA reads
+  default: 10
+  services:
+  - mon
+  - mds
+  - osd
+  - mgr
+  with_legacy: true
+- name: ms_async_ucx_zerocopy
+  type: bool
+  level: advanced
+  default: false
+  with_legacy: true
+- name: ms_async_ucx_feature
+  type: str
+  level: advanced
+  default: am
+  with_legacy: true
+- name: ms_async_ucx_event_polling
+  type: bool
+  level: advanced
+  default: false
+  with_legacy: true
+- name: ms_async_ucx_event_count
+  type: int
+  level: advanced
+  default: 3
+  with_legacy: true
 - name: ms_dpdk_port_id
   type: int
   level: advanced
diff --git a/src/include/config-h.in.cmake b/src/include/config-h.in.cmake
index 6834c8a1a..18421b3ca 100644
--- a/src/include/config-h.in.cmake
+++ b/src/include/config-h.in.cmake
@@ -127,6 +127,9 @@
 /* AsyncMessenger RDMA conditional compilation */
 #cmakedefine HAVE_RDMA
 
+/* AsyncMessenger UCX conditional compilation */
+#cmakedefine HAVE_UCX
+
 /* ibverbs experimental conditional compilation */
 #cmakedefine HAVE_IBV_EXP
 
diff --git a/src/mon/ConfigMonitor.cc b/src/mon/ConfigMonitor.cc
index e24ccbc18..85d5a920a 100644
--- a/src/mon/ConfigMonitor.cc
+++ b/src/mon/ConfigMonitor.cc
@@ -459,6 +459,38 @@ bool ConfigMonitor::preprocess_command(MonOpRequestRef op)
 	}
       }
     }
+
+    // add the lines of ms_async_ucx_tls and ms_async_ucx_device line
+    EntityName entity;
+    entity.from_str("mon.");
+    map<string,string> crush_location;
+    string device_class;
+    std::map<std::string,pair<std::string,const MaskedOption*>> src;
+    auto config = config_map.generate_entity_map(
+      entity,
+      crush_location,
+      mon.osdmon()->osdmap.crush.get(),
+      device_class,
+      &src);
+    
+    string name = ConfFile::normalize_key_name("ms_async_ucx_tls");
+    //const Option *opt = g_conf().find_option(name);
+    auto p = config.find(name);
+    if (p != config.end()) {
+      conf << "\tms_async_ucx_tls = " << p->second << "\n";
+    } else {
+      ss << "unrecognized ms_async_ucx_tls";
+    }
+
+    string name_dev = ConfFile::normalize_key_name("ms_async_ucx_device");
+    //const Option *opt = g_conf().find_option(name);
+    auto q = config.find(name_dev);
+    if (q != config.end()) {
+      conf << "\tms_async_ucx_device = " << q->second << "\n";
+    } else {
+      ss << "unrecognized ms_async_ucx_device";
+    }
+
     odata.append(conf.str());
     err = 0;
   } else {
diff --git a/src/msg/CMakeLists.txt b/src/msg/CMakeLists.txt
index 2aa80c58b..348f4cda6 100644
--- a/src/msg/CMakeLists.txt
+++ b/src/msg/CMakeLists.txt
@@ -39,6 +39,13 @@ if(HAVE_RDMA)
     async/rdma/RDMAStack.cc)
 endif()
 
+if(HAVE_UCX)
+  list(APPEND msg_srcs
+    async/dpdk/UserspaceEvent.cc
+    async/ucx/UCXStack.cc
+    async/ucx/UCXEvent.cc)
+endif(HAVE_UCX)
+
 add_library(common-msg-objs OBJECT ${msg_srcs})
 target_compile_definitions(common-msg-objs PRIVATE
   $<TARGET_PROPERTY:fmt::fmt,INTERFACE_COMPILE_DEFINITIONS>)
diff --git a/src/msg/async/AsyncConnection.cc b/src/msg/async/AsyncConnection.cc
index 5769c580e..944f3797e 100644
--- a/src/msg/async/AsyncConnection.cc
+++ b/src/msg/async/AsyncConnection.cc
@@ -194,6 +194,57 @@ ssize_t AsyncConnection::read(unsigned len, char *buffer,
   return r;
 }
 
+ssize_t AsyncConnection::zero_copy_read_bulk(ceph::bufferlist &bl, size_t length)
+{
+  ssize_t nread;
+  ssize_t total = 0;
+  do {
+    nread = cs.zero_copy_read(bl, length);
+    if (nread < 0) {
+      if (nread == -EAGAIN) {
+        nread = 0;
+      } else if (nread == -EINTR) {
+        nread = 1;// just read again
+      } else {
+        ldout(async_msgr->cct, 1) << __func__ << " reading from fd=" << cs.fd()
+                            << " : "<< nread << " " << strerror(nread) << dendl;
+        return -1;
+      }
+    } else if (nread == 0) {
+      ldout(async_msgr->cct, 1) << __func__ << " peer close file descriptor "
+                                << cs.fd() << dendl;
+      return -1;
+    } else {
+      total += nread;
+      if (bl.length() == length)
+       break;
+    }
+  } while(nread > 0);
+  ldout(async_msgr->cct, 20) << __func__
+                             << " read length " << total
+                             << " bl.length " << bl.length()
+                             << " expected length "<< length << dendl;    
+  return 0;
+}
+
+ssize_t AsyncConnection::zero_copy_read(ceph::bufferlist &bl, size_t length,
+                              std::function<void(ceph::bufferlist &bl, ssize_t)> callback) {
+  ldout(async_msgr->cct, 20) << __func__
+                             << (pendingReadLen ? " continue" : " start")
+                             << " len=" << length << dendl;                             
+  ssize_t r = zero_copy_read_bulk(bl, length);
+  if (r < 0)
+    return r;
+  if ((bl.length() < length)) {
+    zeroCopyReadCallback = callback;
+    pendingReadLen = length;
+    r = length - bl.length();
+    read_bl.swap(bl);
+    return r;
+  }
+  return 0;
+}
+
 // Because this func will be called multi times to populate
 // the needed buffer, so the passed in bufferptr must be the same.
 // Normally, only "read_message" will pass existing bufferptr in
@@ -446,12 +497,20 @@ void AsyncConnection::process() {
 
     case STATE_CONNECTION_ESTABLISHED: {
       if (pendingReadLen) {
-        ssize_t r = read(*pendingReadLen, read_buffer, readCallback);
-        if (r <= 0) { // read all bytes, or an error occured
-          pendingReadLen.reset();
-          char *buf_tmp = read_buffer;
-          read_buffer = nullptr;
-          readCallback(buf_tmp, r);
+        if (read_buffer) {
+          ssize_t r = read(*pendingReadLen, read_buffer, readCallback);
+          if (r <= 0) { // read all bytes, or an error occured
+            pendingReadLen.reset();
+            char *buf_tmp = read_buffer;
+            read_buffer = nullptr;
+            readCallback(buf_tmp, r);
+          }
+        } else {
+          ssize_t r = zero_copy_read_bulk(read_bl, *pendingReadLen);
+          if ((read_bl.length() == *pendingReadLen) || (r < 0)) { // read all bytes, or an error occured
+            pendingReadLen.reset();
+            zeroCopyReadCallback(read_bl, r);
+          }
         }
 	logger->tinc(l_msgr_running_recv_time,
 	    ceph::mono_clock::now() - recv_start_time);
diff --git a/src/msg/async/AsyncConnection.h b/src/msg/async/AsyncConnection.h
index c7f0f9fe8..930f6f764 100644
--- a/src/msg/async/AsyncConnection.h
+++ b/src/msg/async/AsyncConnection.h
@@ -55,6 +55,9 @@ class AsyncConnection : public Connection {
                std::function<void(char *, ssize_t)> callback);
   ssize_t read_until(unsigned needed, char *p);
   ssize_t read_bulk(char *buf, unsigned len);
+  ssize_t zero_copy_read(ceph::bufferlist &bl, size_t length,
+                         std::function<void(ceph::bufferlist &bl, ssize_t)> callback); 
+  ssize_t zero_copy_read_bulk(ceph::bufferlist &bl, size_t length);
 
   ssize_t write(ceph::buffer::list &bl, std::function<void(ssize_t)> callback,
                 bool more=false);
@@ -223,7 +226,9 @@ private:
 
   std::optional<std::function<void(ssize_t)>> writeCallback;
   std::function<void(char *, ssize_t)> readCallback;
+  std::function<void(ceph::bufferlist &, ssize_t)> zeroCopyReadCallback;
   std::optional<unsigned> pendingReadLen;
+  ceph::bufferlist read_bl;
   char *read_buffer;
 
  public:
diff --git a/src/msg/async/AsyncMessenger.cc b/src/msg/async/AsyncMessenger.cc
index 3971e91ff..3cf2425a0 100644
--- a/src/msg/async/AsyncMessenger.cc
+++ b/src/msg/async/AsyncMessenger.cc
@@ -288,6 +288,8 @@ AsyncMessenger::AsyncMessenger(CephContext *cct, entity_name_t name,
     transport_type = "rdma";
   else if (type.find("dpdk") != std::string::npos)
     transport_type = "dpdk";
+  else if (type.find("ucx") != std::string::npos)
+    transport_type = "ucx";
 
   auto single = &cct->lookup_or_create_singleton_object<StackSingleton>(
     "AsyncMessenger::NetworkStack::" + transport_type, true, cct);
diff --git a/src/msg/async/Event.cc b/src/msg/async/Event.cc
index 2c545c07b..10ce34d91 100644
--- a/src/msg/async/Event.cc
+++ b/src/msg/async/Event.cc
@@ -22,6 +22,10 @@
 #include "dpdk/EventDPDK.h"
 #endif
 
+#ifdef HAVE_UCX
+#include "ucx/UCXEvent.h"
+#endif
+
 #ifdef HAVE_EPOLL
 #include "EventEpoll.h"
 #else
@@ -115,6 +119,13 @@ int EventCenter::init(int nevent, unsigned center_id, const std::string &type)
   if (type == "dpdk") {
 #ifdef HAVE_DPDK
     driver = new DPDKDriver(cct);
+#endif
+  } else if (type == "ucx") {
+#ifdef HAVE_UCX
+    ldout(cct, 20) << __func__ << " UCX driver is going to be used" << dendl;
+    driver = new UCXEventDriver(cct);
+#else
+    lderr(cct) << __func__ << " failed to create UCX event driver" << dendl;
 #endif
   } else {
 #ifdef HAVE_EPOLL
diff --git a/src/msg/async/PosixStack.cc b/src/msg/async/PosixStack.cc
index a38e82cf3..57b08ddbf 100644
--- a/src/msg/async/PosixStack.cc
+++ b/src/msg/async/PosixStack.cc
@@ -63,6 +63,10 @@ class PosixConnectedSocketImpl final : public ConnectedSocketImpl {
     }
   }
 
+  ssize_t zero_copy_read(ceph::buffer::list &bl, size_t len) override {
+    return -EOPNOTSUPP;
+  }
+
   ssize_t read(char *buf, size_t len) override {
     #ifdef _WIN32
     ssize_t r = ::recv(_fd, buf, len, 0);
diff --git a/src/msg/async/PosixStack.h b/src/msg/async/PosixStack.h
index a4ff65b28..57a77430e 100644
--- a/src/msg/async/PosixStack.h
+++ b/src/msg/async/PosixStack.h
@@ -47,7 +47,7 @@ class PosixNetworkStack : public NetworkStack {
  public:
   explicit PosixNetworkStack(CephContext *c);
 
-  void spawn_worker(std::function<void ()> &&func) override {
+  void spawn_worker(Worker* worker, std::function<void ()> &&func) override {
     threads.emplace_back(std::move(func));
   }
   void join_worker(unsigned i) override {
diff --git a/src/msg/async/Protocol.h b/src/msg/async/Protocol.h
index 10436307e..a0cc547bd 100644
--- a/src/msg/async/Protocol.h
+++ b/src/msg/async/Protocol.h
@@ -46,8 +46,7 @@ public:
   }
 };
 
-using rx_buffer_t =
-  std::unique_ptr<ceph::buffer::ptr_node, ceph::buffer::ptr_node::disposer>;
+using rx_buffer_t = ceph::buffer::list;
 
 template <class C>
 class CtRxNode : public Ct<C> {
diff --git a/src/msg/async/ProtocolV1.cc b/src/msg/async/ProtocolV1.cc
index 1bc661a55..6ace29047 100644
--- a/src/msg/async/ProtocolV1.cc
+++ b/src/msg/async/ProtocolV1.cc
@@ -1419,7 +1419,7 @@ CtPtr ProtocolV1::handle_server_banner_and_identify(char *buffer, int r) {
       messenger->get_myaddrs().front().is_blank_ip()) {
     sockaddr_storage ss;
     socklen_t len = sizeof(ss);
-    getsockname(connection->cs.fd(), (sockaddr *)&ss, &len);
+    connection->cs.getsockname(connection->cs.fd(), (sockaddr *)&ss, &len);
     entity_addr_t a;
     if (cct->_conf->ms_learn_addr_from_peer) {
       ldout(cct, 1) << __func__ << " peer " << connection->target_addr
diff --git a/src/msg/async/ProtocolV2.cc b/src/msg/async/ProtocolV2.cc
index d63621438..c756b9991 100644
--- a/src/msg/async/ProtocolV2.cc
+++ b/src/msg/async/ProtocolV2.cc
@@ -61,7 +61,7 @@ void ProtocolV2::run_continuation(CtRef continuation) {
 
 #define WRITE(B, D, C) write(D, CONTINUATION(C), B)
 
-#define READ(L, C) read(CONTINUATION(C), ceph::buffer::ptr_node::create(ceph::buffer::create(L)))
+#define READ(L, C) read(CONTINUATION(C), std::move(buffer::ptr(L)))
 
 #define READ_RXBUF(B, C) read(CONTINUATION(C), B)
 
@@ -102,6 +102,9 @@ ProtocolV2::ProtocolV2(AsyncConnection *connection)
                    &session_compression_handlers),
       next_tag(static_cast<Tag>(0)),
       keepalive(false) {
+        if (cct->_conf.get_val<bool>("ms_async_ucx_zerocopy"))
+          use_zero_copy = messenger->get_stack()->support_zero_copy_read();
+        else use_zero_copy = false;  
 }
 
 ProtocolV2::~ProtocolV2() {
@@ -754,14 +757,16 @@ bool ProtocolV2::is_queued() {
 }
 
 CtPtr ProtocolV2::read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next,
-                       rx_buffer_t &&buffer) {
-  const auto len = buffer->length();
-  const auto buf = buffer->c_str();
+                       buffer::ptr &&ptr) {
+  rx_buffer_t buffer;
+  buffer.append(ptr);
+  const auto len = buffer.length();
+  const auto buf = buffer.c_str();
   next.node = std::move(buffer);
   ssize_t r = connection->read(len, buf,
     [&next, this](char *buffer, int r) {
       if (unlikely(pre_auth.enabled) && r >= 0) {
-        pre_auth.rxbuf.append(*next.node);
+        pre_auth.rxbuf.append(next.node);
 	ceph_assert(!cct->_conf->ms_die_on_bug ||
 		    pre_auth.rxbuf.length() < 20000000);
       }
@@ -771,7 +776,7 @@ CtPtr ProtocolV2::read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next,
   if (r <= 0) {
     // error or done synchronously
     if (unlikely(pre_auth.enabled) && r == 0) {
-      pre_auth.rxbuf.append(*next.node);
+      pre_auth.rxbuf.append(next.node);
       ceph_assert(!cct->_conf->ms_die_on_bug ||
 		  pre_auth.rxbuf.length() < 20000000);
     }
@@ -782,6 +787,36 @@ CtPtr ProtocolV2::read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next,
   return nullptr;
 }
 
+CtPtr ProtocolV2::zero_copy_read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next, size_t length) {
+  rx_buffer_t bl;                        
+  ssize_t r = connection->zero_copy_read(bl, length,
+    [&next, this](ceph::bufferlist &bl, int r) {
+      if (unlikely(pre_auth.enabled) && r >= 0) {
+        pre_auth.rxbuf.append(bl);
+	ceph_assert(!cct->_conf->ms_die_on_bug ||
+		    pre_auth.rxbuf.length() < 20000000);
+      }
+           
+      next.node = std::move(bl);
+      next.r = r;
+      run_continuation(next);
+    });
+  if (r <= 0) {
+    // error or done synchronously
+    if (unlikely(pre_auth.enabled) && r == 0) {
+      pre_auth.rxbuf.append(bl);
+      ceph_assert(!cct->_conf->ms_die_on_bug ||
+		  pre_auth.rxbuf.length() < 20000000);
+    }
+
+    next.node = std::move(bl);
+    next.r = r;
+    return &next;
+  }
+
+  return nullptr;
+} 
+
 template <class F>
 CtPtr ProtocolV2::write(const std::string &desc,
                         CONTINUATION_TYPE<ProtocolV2> &next,
@@ -852,7 +887,11 @@ CtPtr ProtocolV2::_banner_exchange(CtRef callback) {
 
 CtPtr ProtocolV2::_wait_for_peer_banner() {
   unsigned banner_len = strlen(CEPH_BANNER_V2_PREFIX) + sizeof(ceph_le16);
-  return READ(banner_len, _handle_peer_banner);
+  if (!use_zero_copy) {
+    return READ(banner_len, _handle_peer_banner); 
+  } else { 
+    return  zero_copy_read(CONTINUATION(_handle_peer_banner), banner_len);      
+  }
 }
 
 CtPtr ProtocolV2::_handle_peer_banner(rx_buffer_t &&buffer, int r) {
@@ -866,8 +905,8 @@ CtPtr ProtocolV2::_handle_peer_banner(rx_buffer_t &&buffer, int r) {
 
   unsigned banner_prefix_len = strlen(CEPH_BANNER_V2_PREFIX);
 
-  if (memcmp(buffer->c_str(), CEPH_BANNER_V2_PREFIX, banner_prefix_len)) {
-    if (memcmp(buffer->c_str(), CEPH_BANNER, strlen(CEPH_BANNER)) == 0) {
+  if (memcmp(buffer.c_str(), CEPH_BANNER_V2_PREFIX, banner_prefix_len)) {
+    if (memcmp(buffer.c_str(), CEPH_BANNER, strlen(CEPH_BANNER)) == 0) {
       lderr(cct) << __func__ << " peer " << *connection->peer_addrs
                  << " is using msgr V1 protocol" << dendl;
       return _fault();
@@ -878,9 +917,7 @@ CtPtr ProtocolV2::_handle_peer_banner(rx_buffer_t &&buffer, int r) {
 
   uint16_t payload_len;
   ceph::bufferlist bl;
-  buffer->set_offset(banner_prefix_len);
-  buffer->set_length(sizeof(ceph_le16));
-  bl.push_back(std::move(buffer));
+  buffer.splice(banner_prefix_len, sizeof(ceph_le16), &bl);
   auto ti = bl.cbegin();
   using ceph::decode;
   try {
@@ -892,7 +929,11 @@ CtPtr ProtocolV2::_handle_peer_banner(rx_buffer_t &&buffer, int r) {
 
   INTERCEPT(state == BANNER_CONNECTING ? 5 : 6);
 
-  return READ(payload_len, _handle_peer_banner_payload);
+  if (!use_zero_copy) {
+    return READ(payload_len, _handle_peer_banner_payload);
+  } else {
+    return  zero_copy_read(CONTINUATION(_handle_peer_banner_payload), payload_len); 
+  }
 }
 
 CtPtr ProtocolV2::_handle_peer_banner_payload(rx_buffer_t &&buffer, int r) {
@@ -909,7 +950,7 @@ CtPtr ProtocolV2::_handle_peer_banner_payload(rx_buffer_t &&buffer, int r) {
 
   ceph::bufferlist bl;
   using ceph::decode;
-  bl.push_back(std::move(buffer));
+  bl.swap(buffer);
   auto ti = bl.cbegin();
   try {
     decode(peer_supported_features, ti);
@@ -990,7 +1031,7 @@ CtPtr ProtocolV2::handle_hello(ceph::bufferlist &payload)
 
   sockaddr_storage ss;
   socklen_t len = sizeof(ss);
-  getsockname(connection->cs.fd(), (sockaddr *)&ss, &len);
+  connection->cs.getsockname(connection->cs.fd(), (sockaddr *)&ss, &len);
   ldout(cct, 5) << __func__ << " getsockname says I am " << (sockaddr *)&ss
 		<< " when talking to " << connection->target_addr << dendl;
 
@@ -1074,8 +1115,12 @@ CtPtr ProtocolV2::read_frame() {
   rx_epilogue.clear();
   rx_segments_data.clear();
 
-  return READ(rx_frame_asm.get_preamble_onwire_len(),
-              handle_read_frame_preamble_main);
+  if (!use_zero_copy) {
+     return READ(rx_frame_asm.get_preamble_onwire_len(),
+               handle_read_frame_preamble_main);
+  } else {
+   return  zero_copy_read(CONTINUATION(handle_read_frame_preamble_main), rx_frame_asm.get_preamble_onwire_len());
+  }
 }
 
 CtPtr ProtocolV2::handle_read_frame_preamble_main(rx_buffer_t &&buffer, int r) {
@@ -1087,7 +1132,7 @@ CtPtr ProtocolV2::handle_read_frame_preamble_main(rx_buffer_t &&buffer, int r) {
     return _fault();
   }
 
-  rx_preamble.push_back(std::move(buffer));
+  rx_preamble.append(std::move(buffer));
 
   ldout(cct, 30) << __func__ << " preamble\n";
   rx_preamble.hexdump(*_dout);
@@ -1100,7 +1145,7 @@ CtPtr ProtocolV2::handle_read_frame_preamble_main(rx_buffer_t &&buffer, int r) {
     return _fault();
   } catch (ceph::crypto::onwire::MsgAuthError&) {
     ldout(cct, 1) << __func__ << "bad auth tag" << dendl;
-    return _fault();
+    return _fault();  
   }
 
   ldout(cct, 25) << __func__ << " disassembled preamble " << rx_frame_asm
@@ -1169,28 +1214,32 @@ CtPtr ProtocolV2::handle_read_frame_dispatch() {
 CtPtr ProtocolV2::read_frame_segment() {
   size_t seg_idx = rx_segments_data.size();
   ldout(cct, 20) << __func__ << " seg_idx=" << seg_idx << dendl;
-  rx_segments_data.emplace_back();
 
   uint32_t onwire_len = rx_frame_asm.get_segment_onwire_len(seg_idx);
   if (onwire_len == 0) {
+    rx_segments_data.emplace_back();
     return _handle_read_frame_segment();
   }
 
-  rx_buffer_t rx_buffer;
   uint16_t align = rx_frame_asm.get_segment_align(seg_idx);
-  try {
-    rx_buffer = ceph::buffer::ptr_node::create(ceph::buffer::create_aligned(
-        onwire_len, align));
-  } catch (const ceph::buffer::bad_alloc&) {
-    // Catching because of potential issues with satisfying alignment.
-    ldout(cct, 1) << __func__ << " can't allocate aligned rx_buffer"
-                  << " len=" << onwire_len
-                  << " align=" << align
-                  << dendl;
-    return _fault();
-  }
 
-  return READ_RXBUF(std::move(rx_buffer), handle_read_frame_segment);
+  if (!use_zero_copy) {
+    ceph::unique_leakable_ptr<buffer::raw> raw_buf;
+    try {
+      raw_buf = ceph::buffer::create_aligned(onwire_len, align);
+    } catch (std::bad_alloc&) {
+      // Catching because of potential issues with satisfying alignment.
+      ldout(cct, 20) << __func__ << " can't allocate aligned raw_buf "
+    	   << " len=" << onwire_len
+    	   << " align=" << align
+    	   << dendl;
+      return _fault();
+    }
+
+    return read(CONTINUATION(handle_read_frame_segment), std::move(bufferptr (std::move(raw_buf))));
+  } else {
+    return  zero_copy_read(CONTINUATION(handle_read_frame_segment), onwire_len);
+  }
 }
 
 CtPtr ProtocolV2::handle_read_frame_segment(rx_buffer_t &&rx_buffer, int r) {
@@ -1202,7 +1251,14 @@ CtPtr ProtocolV2::handle_read_frame_segment(rx_buffer_t &&rx_buffer, int r) {
     return _fault();
   }
 
-  rx_segments_data.back().push_back(std::move(rx_buffer));
+  size_t seg_idx = rx_segments_data.size();
+  uint16_t align = rx_frame_asm.get_segment_align(seg_idx);
+  if (!rx_buffer.is_aligned(align)) {
+    rx_buffer.rebuild_aligned_size_and_memory(align, align);
+  }
+
+  rx_segments_data.emplace_back();
+  rx_segments_data.back().append(std::move(rx_buffer));
   return _handle_read_frame_segment();
 }
 
@@ -1213,7 +1269,12 @@ CtPtr ProtocolV2::_handle_read_frame_segment() {
     if (epilogue_onwire_len == 0) {
       return _handle_read_frame_epilogue_main();
     }
-    return READ(epilogue_onwire_len, handle_read_frame_epilogue_main);
+    
+    if (!use_zero_copy) {
+      return READ(epilogue_onwire_len, handle_read_frame_epilogue_main);
+    } else {
+      return  zero_copy_read(CONTINUATION(handle_read_frame_epilogue_main), epilogue_onwire_len);
+    }
   }
   // TODO: for makeshift only. This will be more generic and throttled
   return read_frame_segment();
@@ -1320,7 +1381,7 @@ CtPtr ProtocolV2::handle_read_frame_epilogue_main(rx_buffer_t &&buffer, int r)
     return _fault();
   }
 
-  rx_epilogue.push_back(std::move(buffer));
+  rx_epilogue.append(std::move(buffer));
   return _handle_read_frame_epilogue_main();
 }
 
@@ -3014,3 +3075,4 @@ CtPtr ProtocolV2::finish_compression() {
   state = SESSION_ACCEPTING;
   return CONTINUE(read_frame);
 }
+
diff --git a/src/msg/async/ProtocolV2.h b/src/msg/async/ProtocolV2.h
index 6441866fe..f9d7c8eb5 100644
--- a/src/msg/async/ProtocolV2.h
+++ b/src/msg/async/ProtocolV2.h
@@ -122,13 +122,23 @@ private:
   bool keepalive;
   bool write_in_progress = false;
 
+  bool use_zero_copy;
+  // stat counters
+  enum {
+    l_msgr_protocalv2_first = 98000,
+    l_msgr_protocalv2_zerocopy,
+    l_msgr_protocalv2_copy,
+    l_msgr_protocalv2_last,
+  };
+
   CompConnectionMeta comp_meta;
   std::ostream& _conn_prefix(std::ostream *_dout);
   void run_continuation(Ct<ProtocolV2> *pcontinuation);
   void run_continuation(Ct<ProtocolV2> &continuation);
 
   Ct<ProtocolV2> *read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next,
-                       rx_buffer_t&& buffer);
+                       buffer::ptr &&ptr);
+  Ct<ProtocolV2> *zero_copy_read(CONTINUATION_RXBPTR_TYPE<ProtocolV2> &next, size_t length);
   template <class F>
   Ct<ProtocolV2> *write(const std::string &desc,
                         CONTINUATION_TYPE<ProtocolV2> &next,
diff --git a/src/msg/async/Stack.cc b/src/msg/async/Stack.cc
index 94a1bba53..661709444 100644
--- a/src/msg/async/Stack.cc
+++ b/src/msg/async/Stack.cc
@@ -23,6 +23,9 @@
 #ifdef HAVE_RDMA
 #include "rdma/RDMAStack.h"
 #endif
+#ifdef HAVE_UCX
+#include "ucx/UCXStack.h"
+#endif
 #ifdef HAVE_DPDK
 #include "dpdk/DPDKStack.h"
 #endif
@@ -75,6 +78,10 @@ std::shared_ptr<NetworkStack> NetworkStack::create(CephContext *c,
   else if (t == "dpdk")
     stack.reset(new DPDKStack(c));
 #endif
+#ifdef HAVE_UCX
+  else if (t == "ucx")
+    stack.reset(new UCXStack(c));
+#endif
 
   if (stack == nullptr) {
     lderr(c) << __func__ << " ms_async_transport_type " << t <<
@@ -119,7 +126,7 @@ void NetworkStack::start()
   for (Worker* worker : workers) {
     if (worker->is_init())
       continue;
-    spawn_worker(add_thread(worker));
+    spawn_worker(worker, add_thread(worker));
   }
   started = true;
   lk.unlock();
diff --git a/src/msg/async/Stack.h b/src/msg/async/Stack.h
index 97201787c..6f1301029 100644
--- a/src/msg/async/Stack.h
+++ b/src/msg/async/Stack.h
@@ -28,10 +28,14 @@ class ConnectedSocketImpl {
   virtual ~ConnectedSocketImpl() {}
   virtual int is_connected() = 0;
   virtual ssize_t read(char*, size_t) = 0;
+  virtual ssize_t zero_copy_read(ceph::buffer::list &bl, size_t len) = 0;
   virtual ssize_t send(ceph::buffer::list &bl, bool more) = 0;
   virtual void shutdown() = 0;
   virtual void close() = 0;
   virtual int fd() const = 0;
+  virtual int getsockname(int sockfd, struct sockaddr * addr, socklen_t * addrlen){
+	  return ::getsockname(sockfd, addr, addrlen);
+  }
 };
 
 class ConnectedSocket;
@@ -93,6 +97,9 @@ class ConnectedSocket {
   ssize_t read(char* buf, size_t len) {
     return _csi->read(buf, len);
   }
+  ssize_t zero_copy_read(ceph::buffer::list &bl, size_t len) {
+    return _csi->zero_copy_read(bl, len);
+  }
   /// Gets the output stream.
   ///
   /// Gets an object that sends data to the remote endpoint.
@@ -122,7 +129,10 @@ class ConnectedSocket {
   int fd() const {
     return _csi->fd();
   }
-
+  int getsockname(int sockfd, struct sockaddr *addr,
+                  socklen_t * addrlen) {
+      return _csi->getsockname(sockfd, addr, addrlen);
+  }
   explicit operator bool() const {
     return _csi.get();
   }
@@ -305,6 +315,8 @@ class NetworkStack {
     sprintf(tp_name, "msgr-worker-%u", id);
     ceph_pthread_setname(pthread_self(), tp_name);
   }
+ public:
+  virtual bool support_zero_copy_read() const { return false; }
 
  protected:
   CephContext *cct;
@@ -343,7 +355,7 @@ class NetworkStack {
   }
 
   // direct is used in tests only
-  virtual void spawn_worker(std::function<void ()> &&) = 0;
+  virtual void spawn_worker(Worker* worker, std::function<void ()> &&) = 0;
   virtual void join_worker(unsigned i) = 0;
 
   virtual bool is_ready() { return true; };
diff --git a/src/msg/async/dpdk/DPDKStack.cc b/src/msg/async/dpdk/DPDKStack.cc
index 1543a530d..c954de710 100644
--- a/src/msg/async/dpdk/DPDKStack.cc
+++ b/src/msg/async/dpdk/DPDKStack.cc
@@ -243,7 +243,7 @@ int DPDKWorker::connect(const entity_addr_t &addr, const SocketOptions &opts, Co
   return r;
 }
 
-void DPDKStack::spawn_worker(std::function<void ()> &&func)
+void DPDKStack::spawn_worker(Worker* worker, std::function<void ()> &&func)
 {
   // create a extra master thread
   //
diff --git a/src/msg/async/dpdk/DPDKStack.h b/src/msg/async/dpdk/DPDKStack.h
index f05873572..fad0ae6ea 100644
--- a/src/msg/async/dpdk/DPDKStack.h
+++ b/src/msg/async/dpdk/DPDKStack.h
@@ -260,9 +260,10 @@ class DPDKStack : public NetworkStack {
   explicit DPDKStack(CephContext *cct): NetworkStack(cct), eal(cct) {
     funcs.reserve(cct->_conf->ms_async_op_threads);
   }
+  virtual bool support_zero_copy_read() const override { return true; }
   virtual bool support_local_listen_table() const override { return true; }
 
-  virtual void spawn_worker(std::function<void ()> &&func) override;
+  virtual void spawn_worker(Worker* worker, std::function<void ()> &&func) override;
   virtual void join_worker(unsigned i) override;
  private:
   dpdk::eal eal;
diff --git a/src/msg/async/dpdk/UserspaceEvent.cc b/src/msg/async/dpdk/UserspaceEvent.cc
index e0c57fd9b..117af1831 100644
--- a/src/msg/async/dpdk/UserspaceEvent.cc
+++ b/src/msg/async/dpdk/UserspaceEvent.cc
@@ -110,6 +110,8 @@ int UserspaceEventManager::poll(int *events, int *masks, int num_events, struct
 
     events[count] = fd;
     std::optional<UserspaceFDImpl> &impl = fds[fd];
+    if (impl->waiting_idx == 0)
+      continue;
     ceph_assert(impl);
     masks[count] = impl->listening_mask & impl->activating_mask;
     ceph_assert(masks[count]);
diff --git a/src/msg/async/dpdk/UserspaceEvent.h b/src/msg/async/dpdk/UserspaceEvent.h
index 49308aca4..2c21c1247 100644
--- a/src/msg/async/dpdk/UserspaceEvent.h
+++ b/src/msg/async/dpdk/UserspaceEvent.h
@@ -30,6 +30,8 @@
 class CephContext;
 
 class UserspaceEventManager {
+  std::list<uint32_t> unused_fds;
+protected:
   struct UserspaceFDImpl {
     uint32_t waiting_idx = 0;
     int16_t read_errno = 0;
@@ -43,7 +45,6 @@ class UserspaceEventManager {
   uint32_t max_wait_idx = 0;
   std::vector<std::optional<UserspaceFDImpl> > fds;
   std::vector<int> waiting_fds;
-  std::list<uint32_t> unused_fds;
 
  public:
   explicit UserspaceEventManager(CephContext *c): cct(c) {
diff --git a/src/msg/async/rdma/RDMAConnectedSocketImpl.cc b/src/msg/async/rdma/RDMAConnectedSocketImpl.cc
index 5ab6c9b2e..97c92c027 100644
--- a/src/msg/async/rdma/RDMAConnectedSocketImpl.cc
+++ b/src/msg/async/rdma/RDMAConnectedSocketImpl.cc
@@ -341,6 +341,11 @@ ssize_t RDMAConnectedSocketImpl::read_buffers(char* buf, size_t len)
   return read_size;
 }
 
+ssize_t RDMAConnectedSocketImpl::zero_copy_read(ceph::buffer::list &bl, size_t len)
+{
+  return -EOPNOTSUPP;
+}
+
 ssize_t RDMAConnectedSocketImpl::send(ceph::buffer::list &bl, bool more)
 {
   if (error) {
diff --git a/src/msg/async/rdma/RDMAStack.cc b/src/msg/async/rdma/RDMAStack.cc
index 12db599d6..fc1cb30de 100644
--- a/src/msg/async/rdma/RDMAStack.cc
+++ b/src/msg/async/rdma/RDMAStack.cc
@@ -800,7 +800,7 @@ Worker* RDMAStack::create_worker(CephContext *c, unsigned worker_id)
   return w;
 }
 
-void RDMAStack::spawn_worker(std::function<void ()> &&func)
+void RDMAStack::spawn_worker(Worker* worker, std::function<void ()> &&func)
 {
   threads.emplace_back(std::move(func));
 }
diff --git a/src/msg/async/rdma/RDMAStack.h b/src/msg/async/rdma/RDMAStack.h
index aa64ca57e..cb250ce92 100644
--- a/src/msg/async/rdma/RDMAStack.h
+++ b/src/msg/async/rdma/RDMAStack.h
@@ -214,6 +214,7 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
   virtual int is_connected() override { return connected; }
 
   virtual ssize_t read(char* buf, size_t len) override;
+  virtual ssize_t zero_copy_read(ceph::buffer::list &bl, size_t len) override;
   virtual ssize_t send(ceph::buffer::list &bl, bool more) override;
   virtual void shutdown() override;
   virtual void close() override;
@@ -331,9 +332,10 @@ class RDMAStack : public NetworkStack {
  public:
   explicit RDMAStack(CephContext *cct);
   virtual ~RDMAStack();
+  virtual bool support_zero_copy_read() const override { return false; }
   virtual bool nonblock_connect_need_writable_event() const override { return false; }
 
-  virtual void spawn_worker(std::function<void ()> &&func) override;
+  virtual void spawn_worker(Worker* worker, std::function<void ()> &&func) override;
   virtual void join_worker(unsigned i) override;
   virtual bool is_ready() override { return fork_finished.load(); };
   virtual void ready() override { fork_finished = true; };
diff --git a/src/msg/async/ucx/UCXEvent.cc b/src/msg/async/ucx/UCXEvent.cc
new file mode 100644
index 000000000..e07352aa7
--- /dev/null
+++ b/src/msg/async/ucx/UCXEvent.cc
@@ -0,0 +1,153 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (c) Huawei Technologies Co., Ltd. 2023-2023.
+ *
+ * Author: Chunsong Feng <fengchunsong@huawei.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "UCXEvent.h"
+#include "UCXStack.h"
+#include "common/Cycles.h"
+
+#define dout_subsys ceph_subsys_ms
+
+#undef dout_prefix
+#define dout_prefix *_dout << "UCXEventDriver "
+
+int UCXEventManager::get_eventfd()
+{
+    std::lock_guard l(lock);
+    int fd = dup(base_fd);
+    if (fd > max_fd) {
+        max_fd = fd;
+        fds.resize(max_fd + 1);
+    }
+
+    std::optional<UserspaceFDImpl> &impl = fds[fd];
+    ceph_assert(!impl);
+    impl.emplace();
+    return fd;
+}
+
+void UCXEventManager::close(int fd)
+{
+    std::lock_guard l(lock);
+    if ((size_t)fd >= fds.size()) {
+        return;
+    }
+
+    std::optional<UserspaceFDImpl> &impl = fds[fd];
+    if (!impl) {
+        return;
+    }
+
+    if (fd == max_fd) {
+        max_fd--;
+    }
+
+    if (impl->activating_mask) {
+        if (waiting_fds[max_wait_idx] == fd) {
+            ceph_assert(impl->waiting_idx == max_wait_idx);
+            --max_wait_idx;
+        }
+        waiting_fds[impl->waiting_idx] = -1;
+    }
+    impl.reset();
+    ::close(fd);
+    fd = -1;
+}
+
+int UCXEventDriver::init(EventCenter *c, int nevent)
+{
+    ldout(cct, 20) << __func__ << " num of event_masks: " << nevent << dendl;
+    Cycles::init();
+    return EpollDriver::init(c, nevent);
+}
+
+int UCXEventDriver::add_event(int fd, int cur_mask, int add_mask)
+{
+    ldout(cct, 20) << __func__ << " fd: " << fd << " read ? " << (EVENT_READABLE & add_mask) << dendl;
+
+    int r = manager.listen(fd, add_mask);
+    if (r == -ENOENT) {
+        return EpollDriver::add_event(fd, cur_mask, add_mask);
+    }
+
+    return r;
+}
+
+int UCXEventDriver::del_event(int fd, int cur_mask, int delmask)
+{
+    ldout(cct, 20) << __func__ << " del event fd=" << fd << " cur_mask=" << cur_mask << " delmask=" << delmask << dendl;
+    int r = 0;
+
+    if (delmask != EVENT_NONE) {
+        r = manager.unlisten(fd, delmask);
+        if (r == -ENOENT) {
+            return EpollDriver::del_event(fd, cur_mask, delmask);
+        }
+    }
+
+    return r;
+}
+
+int UCXEventDriver::resize_events(int newsize)
+{
+    return EpollDriver::resize_events(newsize);
+}
+
+int UCXEventDriver::event_wait(std::vector<FiredFileEvent> &fired_events, struct timeval *tvp)
+{
+    if (nullptr == ucp_worker) {
+        ldout(cct, 1) << __func__ << " ucp_worker null " << dendl;
+        return EpollDriver::event_wait(fired_events, tvp);
+    }
+
+    uint64_t start = Cycles::rdtsc();
+    tvp->tv_usec = cct->_conf->ms_async_rdma_polling_us;
+    static int count = cct->_conf->ms_async_ucx_event_count;
+    double seconds = tvp->tv_sec + tvp->tv_usec / 1000000.f;
+    uint64_t end = Cycles::from_seconds(seconds) + start;
+    while (ucp_worker_progress(ucp_worker)) {
+        uint64_t now = Cycles::rdtsc();
+        if (now > end) {
+            break;
+        }
+    }
+
+    int retval = manager.poll(ucx_events, ucx_masks, num_events, tvp);
+    if (retval > 0) {
+        fired_events.resize(retval);
+        for (int i = 0; i < retval; i++) {
+            fired_events[i].fd = ucx_events[i];
+            fired_events[i].mask = ucx_masks[i];
+        }
+        count = cct->_conf->ms_async_ucx_event_count;
+        return fired_events.size();
+    }
+    
+    count = (count == 0) ? 0 : count - 1;
+    bool is_polling = cct->_conf->ms_async_ucx_event_polling;
+    if (false == is_polling && 0 == count) {
+        if (UCS_OK == ucp_worker_arm(ucp_worker)) {
+            int retval = EpollDriver::event_wait(fired_events, tvp);
+            if (retval < 0) {
+                lderr(cct) << __func__ << " epoll wait failed with err: " << retval << dendl;
+                return retval;
+            }
+
+            ldout(cct, 20) << __func__ << " waking up with " << fired_events.size() << " events " << dendl;
+        }
+    }
+
+    return fired_events.size();
+}
diff --git a/src/msg/async/ucx/UCXEvent.h b/src/msg/async/ucx/UCXEvent.h
new file mode 100644
index 000000000..21a8256de
--- /dev/null
+++ b/src/msg/async/ucx/UCXEvent.h
@@ -0,0 +1,78 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (c) Huawei Technologies Co., Ltd. 2023-2023.
+ *
+ * Author: Chunsong Feng <fengchunsong@huawei.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef CEPH_UCXEVENT_H
+#define CEPH_UCXEVENT_H
+
+#include "msg/async/Event.h"
+#include "msg/async/EventEpoll.h"
+#include "common/ceph_mutex.h"
+#include "msg/async/dpdk/UserspaceEvent.h"
+
+extern "C" {
+#include <ucp/api/ucp.h>
+};
+
+class UCXEventManager : public UserspaceEventManager {
+    CephContext *cct;
+    ceph::mutex lock = ceph::make_mutex("UCXEventManager::lock");
+    int base_fd;
+
+public:
+    explicit UCXEventManager(CephContext *c) : UserspaceEventManager(c), cct(c) {};
+    virtual ~UCXEventManager() {};
+
+    int get_eventfd();
+    void close(int fd);
+
+    void init(int fd)
+    {
+        base_fd = fd;
+        max_fd = fd;
+    }
+};
+
+class UCXEventDriver : public EpollDriver {
+    UCXEventManager manager;
+
+public:
+    explicit UCXEventDriver(CephContext *c) : EpollDriver(c), manager(c), cct(c) {};
+    virtual ~UCXEventDriver() {};
+
+    int init(EventCenter *c, int nevent) override;
+    int add_event(int fd, int cur_mask, int add_mask) override;
+    int del_event(int fd, int cur_mask, int del_mask) override;
+    int resize_events(int newsize) override;
+    int event_wait(std::vector<FiredFileEvent> &fired_events, struct timeval *tp) override;
+    void driver_init(ucp_worker_h ucp_w, int fd)
+    {
+        ucp_worker = ucp_w;
+        manager.init(fd);
+    }
+
+    UCXEventManager *get_manager() {
+        return &manager;
+    }
+
+private:
+    CephContext *cct;
+    ucp_worker_h ucp_worker;
+    const static int num_events = 512;
+    int ucx_events[num_events];
+    int ucx_masks[num_events];
+};
+
+#endif // CEPH_UCXEVENT_H
diff --git a/src/msg/async/ucx/UCXStack.cc b/src/msg/async/ucx/UCXStack.cc
new file mode 100644
index 000000000..d971956ff
--- /dev/null
+++ b/src/msg/async/ucx/UCXStack.cc
@@ -0,0 +1,1076 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (c) Huawei Technologies Co., Ltd. 2023-2023.
+ *
+ * Author: Chunsong Feng <fengchunsong@huawei.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include "UCXStack.h"
+#include "UCXEvent.h"
+#include "common/deleter.h"
+
+#define dout_subsys ceph_subsys_ms
+#undef dout_prefix
+#define dout_prefix *_dout << "UCXConnectedSocketImpl "
+
+UCXConnectedSocketImpl::UCXConnectedSocketImpl(UCXWorker *w, int fd) : cct(w->cct), worker(w), conn_fd(fd)
+{
+    UCXEventDriver *driver = dynamic_cast<UCXEventDriver *>(worker->center.get_driver());
+    manager = driver->get_manager();
+    ucp_worker = w->get_ucp_worker();
+    use_zero_copy = cct->_conf.get_val<bool>("ms_async_ucx_zerocopy");
+}
+
+void UCXConnectedSocketImpl::shutdown()
+{
+    ceph_assert(worker->center.in_thread());
+    ldout(cct, 20) << __func__ << " conn fd: " << conn_fd << " is shutting down..." << dendl;
+}
+
+void UCXConnectedSocketImpl::request_init(void *req)
+{
+    UCXReqDescr *desc = static_cast<UCXReqDescr *>(req);
+    desc->bl = new bufferlist;
+}
+
+void UCXConnectedSocketImpl::request_cleanup(void *req)
+{
+    UCXReqDescr *desc = static_cast<UCXReqDescr *>(req);
+
+    delete desc->bl;
+}
+
+void UCXConnectedSocketImpl::ucx_ep_close(bool close_event)
+{
+    int fd = conn_fd;
+    if (ucp_ep != nullptr) {
+        ceph_assert(nullptr != ucp_ep);
+        ldout(cct, 20) << __func__ << " fd: " << fd << " ep=" << (void *)ucp_ep << dendl;
+        for (auto req : send_reqs) {
+            req->conn = nullptr;
+            ucp_request_cancel(ucp_worker, req);
+        }
+        send_reqs.clear();
+        for (auto req : recv_reqs) {
+            req->conn = nullptr;
+            ucp_request_cancel(ucp_worker, req);
+        }
+        recv_reqs.clear();
+
+        ucs_status_ptr_t request = ucp_ep_close_nb(ucp_ep, UCP_EP_CLOSE_MODE_FORCE);
+        if (nullptr == request) {
+            ldout(cct, 10) << __func__ << " ucp ep fd: " << fd << " closed in place..." << dendl;
+        } else if (UCS_PTR_IS_PTR(request)) {
+            /*
+             * We don't care even the request is still in
+             * UCS_INPROGRESS state - it's UCX responsility now
+             */
+            ldout(cct, 10) << __func__ << " ucp ep fd: " << fd << " closed pending..." << dendl;
+            ucp_request_free(request);
+        } else if (UCS_PTR_STATUS(request) != UCS_OK) {
+            lderr(cct) << __func__ << " fd: " << fd << " ucp_ep_close_nb call failed: err " <<
+                ucs_status_string(UCS_PTR_STATUS(request)) << dendl;
+        }
+
+        worker->erase_connection(ucp_ep);
+        conn_release_recvs();
+        ucp_ep = nullptr;
+        conn_state = CLOSED;
+    }
+
+    ldout(cct, 10) << __func__ << " fd: " << fd << " exit..." << dendl;
+}
+
+void UCXConnectedSocketImpl::conn_release_recvs()
+{
+    size_t total = rx_bl.length();
+
+    /* Free all pending_fd receives */
+    rx_bl.clear();
+    if (total)
+        ldout(cct, 1) << __func__ << " fd: " << conn_fd << " total " << total << dendl;
+}
+
+void UCXConnectedSocketImpl::close()
+{
+    if (conn_fd > 0) {
+        ldout(cct, 10) << __func__ << " fd: " << conn_fd << " send " << sn_send << " recv " << sn_recv << dendl;
+        ucx_ep_close(false);
+        manager->close(conn_fd);
+        conn_fd = -1;
+    }
+}
+
+UCXConnectedSocketImpl::~UCXConnectedSocketImpl()
+{
+    UCXConnectedSocketImpl::close();
+    out_of_order_map.clear();
+}
+
+int UCXConnectedSocketImpl::submit(bufferlist &bl, MsgHeader *msg, const void *buffer, size_t count,
+    ucp_datatype_t type)
+{
+    ucp_dt_iov_t *iov_list = nullptr;
+    msg->seq = sn_send;
+    msg->hton();
+    ucp_request_param_t param;
+    param.op_attr_mask =
+        UCP_OP_ATTR_FIELD_CALLBACK | UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_USER_DATA | UCP_OP_ATTR_FIELD_FLAGS;
+    param.datatype = type;
+    param.cb.send = send_completion_cb;
+    param.flags = UCP_AM_SEND_FLAG_REPLY;
+    param.user_data = reinterpret_cast<void *>(msg);
+    UCXReqDescr *req =
+        static_cast<UCXReqDescr *>(ucp_am_send_nbx(ucp_ep, AM_ID, msg, sizeof(*msg), buffer, count, &param));
+
+    if ((type & UCP_DATATYPE_CLASS_MASK) == UCP_DATATYPE_IOV)
+        iov_list = (ucp_dt_iov_t *)buffer;
+    if (req == nullptr) {
+        /* in place completion */
+        ldout(cct, 20) << __func__ << " SENT IN PLACE " << dendl;
+        bl.splice(0, msg->data_len);
+        sn_send += msg->data_len;
+        if (iov_list) {
+            delete[] iov_list;
+        }
+        delete msg;
+        return 0;
+    }
+
+    if (UCS_PTR_IS_ERR(req)) {
+        if (iov_list) {
+            delete[] iov_list;
+        }
+        delete msg;
+        lderr(cct) << __func__ << " fd: " << conn_fd << " send failure: " << UCS_PTR_STATUS(req) << dendl;
+        return -EAGAIN;
+    }
+
+    req->conn = this;
+    bl.splice(0, msg->data_len, req->bl);
+    req->iov_list = iov_list;
+    req->sn = sn_send;
+    send_reqs.insert(req);
+    ldout(cct, 20) << __func__ << " send in progress req " << req << dendl;
+
+    sn_send += msg->data_len;
+    return 0;
+}
+
+ssize_t UCXConnectedSocketImpl::do_send(bufferlist &bl)
+{
+    unsigned iov_cnt = bl.get_num_buffers();
+
+    auto pb = std::cbegin(bl.buffers());
+    int rc;
+    size_t length = 0;
+    if (iov_cnt == 1) {
+        MsgHeader *msg = new MsgHeader;
+        length = pb->length();
+        msg->data_len = length;
+        if (pb->length() >= PAGE_SIZE_ALIGNMENT) {
+            msg->align = 0;
+        }
+        rc = submit(bl, msg, pb->c_str(), pb->length(), ucp_dt_make_contig(1));
+        if (rc < 0) {
+            return rc;
+        }
+        return length;
+    } else {
+        ucp_dt_iov_t *iov_list = new ucp_dt_iov_t[iov_cnt];
+        MsgHeader *msg = new MsgHeader;
+        int n;
+        size_t off = 0;
+        const char *pbuf = pb->raw_c_str();
+        int buf_count = 0;
+        for (n = 0; pb != bl.buffers().end(); ++pb) {
+            if (pb->length() == 0) {
+                ldout(cct, 10) << __func__ << " skip zero len " << dendl;
+                continue;
+            }
+
+            iov_list[n].buffer = (void *)(pb->c_str());
+            iov_list[n].length = pb->length();
+
+            if (pb->length() >= PAGE_SIZE_ALIGNMENT) {
+                msg->align = off;
+            }
+            off += pb->length();
+            n++;
+            if (pb->raw_c_str() == pbuf) {
+                buf_count++;
+                if (buf_count == 3) {
+                    buf_count = 0;
+                    break;
+                }
+            }
+        }
+        length = off;
+        msg->data_len = off;
+        rc = submit(bl, msg, iov_list, n, ucp_dt_make_iov());
+        if (rc < 0) {
+            return rc;
+        }
+    }
+
+    return length;
+}
+
+ssize_t UCXConnectedSocketImpl::send(bufferlist &bl, bool more)
+{
+    unsigned total_len = bl.length();
+
+    if (total_len == 0) {
+        ldout(cct, 1) << __func__ << " fd: " << conn_fd << " sending " << total_len << dendl;
+        return 0;
+    }
+
+    if (conn_state >= DISCONNECTED)
+        return -EIO;
+
+    if (nullptr == ucp_ep) {
+        ldout(cct, 20) << __func__ << " fd: " << conn_fd << " put " << total_len << " bytes to the pending" << dendl;
+        return -EIO;
+    }
+    
+    ssize_t rc = do_send(bl);
+    if (rc < 0) {
+        return rc;
+    }
+    if (!more) {
+        while (ucp_worker_progress(ucp_worker))
+            ;
+    }
+
+    return total_len - bl.length();
+}
+
+int UCXConnectedSocketImpl::getsockname(int sockfd, struct sockaddr *addr, socklen_t *addrlen)
+{
+    ucp_ep_attr_t attr;
+    attr.field_mask = UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR;
+    ucs_status_t status = ucp_ep_query(ucp_ep, &attr);
+    if (status != UCS_OK) {
+        return -1;
+    }
+    if (*addrlen < sizeof(struct sockaddr)) {
+        memcpy(addr, (char *)(&attr.local_sockaddr), *addrlen);
+        *addrlen = sizeof(struct sockaddr);
+    } else {
+        memcpy(addr, (char *)(&attr.local_sockaddr), sizeof(struct sockaddr));
+    }
+
+    return 0;
+};
+
+void UCXConnectedSocketImpl::send_completion_cb(void *req, ucs_status_t status, void *userdata)
+{
+    UCXReqDescr *descr = static_cast<UCXReqDescr *>(req);
+    UCXConnectedSocketImpl *conn = descr->conn;
+    MsgHeader *msg = reinterpret_cast<MsgHeader *>(userdata);
+    if (conn) {
+        conn->send_reqs.erase(descr);
+        ldout(conn->cct, 20) << __func__ << " send complete sn_send " << descr->sn << " length " <<
+            descr->bl->length() << dendl;
+    }
+    descr->bl->clear();
+    if (descr->iov_list) {
+        delete[] descr->iov_list;
+    }
+    delete msg;
+
+    ucp_request_free(req);
+}
+
+void UCXConnectedSocketImpl::insert_recv_buffers(uint64_t sn, bufferlist &bl, size_t length)
+{
+    if (sn > sn_recv) {
+        ldout(cct, 20) << __func__ << " out of oder, expected " << sn_recv << " got " << sn << " length " <<
+            bl.length() << dendl;
+        insert_out_of_order(sn, bl);
+    } else if (sn_recv - sn + bl.length() <= length) {
+        sn_recv += bl.length();
+        rx_bl.append(std::move(bl));
+        merge_out_of_order();
+        manager->notify(conn_fd, EVENT_READABLE);
+    } else {
+        ldout(cct, 0) << __func__ << " out of oder error " << sn_recv << " got " << sn << " length " << bl.length() <<
+            " msg length " << length << dendl;
+    }
+}
+
+void UCXConnectedSocketImpl::am_recv_data_callback(void *req, ucs_status_t status, size_t length, void *user_data)
+{
+    UCXReqDescr *desc = reinterpret_cast<UCXReqDescr *>(req);
+    UCXWorker *myworker = reinterpret_cast<UCXWorker *>(user_data);
+    UCXConnectedSocketImpl *conn = desc->conn;
+
+    if (conn == nullptr) {
+        lderr(myworker->cct) << " callback after disconnect status " << status << dendl;
+        goto free_resource;
+    }
+
+    if (status != UCS_OK) {
+        lderr(myworker->cct) << " callback with error status " << ucs_status_string(status) << dendl;
+        goto remove_req;
+    }
+
+    ceph_assert(desc->bl->length() == length);
+    conn->insert_recv_buffers(desc->sn, *desc->bl, desc->bl->length());
+    ldout(myworker->cct, 20) << " am_recv_data_callback with length " << length << dendl;
+
+remove_req:
+    conn->recv_reqs.erase(desc);
+
+free_resource:
+    desc->bl->clear();
+    if (desc->iov_list) {
+        delete[] desc->iov_list;
+    }
+    ucp_request_free(req);
+}
+
+bufferlist UCXConnectedSocketImpl::create_bufferlist_with_hint(size_t length, MsgHeader &msg)
+{
+    bufferlist bl;
+    size_t len = length;
+    if (!use_zero_copy) {
+        bufferptr ptr(buffer::create(length));
+        bl.append(std::move(ptr));
+        return bl;
+    }
+    if (msg.align != 0xFFFFFFFF) {
+        len = msg.align;
+    }
+    if (len) {
+        bufferptr ptr(buffer::create(len));
+        bl.append(std::move(ptr));
+    }
+    len = length - len;
+    if (len) {
+        bufferptr ptr(buffer::create_aligned(len, PAGE_SIZE_ALIGNMENT));
+        bl.append(std::move(ptr));
+    }
+    return bl;
+}
+
+ucs_status_t UCXConnectedSocketImpl::process_recv_request(ucp_worker_h ucp_worker, void *data, size_t length,
+    MsgHeader &msg)
+{
+    bufferlist bl = create_bufferlist_with_hint(length, msg);
+
+    unsigned iov_cnt = bl.get_num_buffers();
+
+    ucp_request_param_t param;
+    param.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK | UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_USER_DATA;
+    param.datatype = (iov_cnt == 1) ? ucp_dt_make_contig(1) : ucp_dt_make_iov();
+    param.cb.recv_am = am_recv_data_callback;
+    param.user_data = reinterpret_cast<void *>(worker);
+    ucp_dt_iov_t *iov_list;
+    UCXReqDescr *req;
+    auto pb = std::cbegin(bl.buffers());
+    if (iov_cnt == 1) {
+        iov_list = nullptr;
+
+        req = static_cast<UCXReqDescr *>(
+            ucp_am_recv_data_nbx(ucp_worker, data, (void *)(pb->c_str()), pb->length(), &param));
+    } else {
+        iov_list = new ucp_dt_iov_t[iov_cnt];
+        for (int n = 0; pb != bl.buffers().end(); ++pb, ++n) {
+            iov_list[n].buffer = (void *)(pb->c_str());
+            iov_list[n].length = pb->length();
+        }
+        req = static_cast<UCXReqDescr *>(ucp_am_recv_data_nbx(ucp_worker, data, iov_list, iov_cnt, &param));
+    }
+    if (req == nullptr) {
+        if (iov_list) {
+            delete[] iov_list;
+        }
+        insert_recv_buffers(msg.seq, bl, msg.data_len);
+        bl.clear();
+        lderr(worker->cct) << __func__ << " read inplace length " << length << " total " << sn_recv << dendl;
+        return UCS_OK;
+    } else if (UCS_PTR_IS_ERR(req)) {
+        if (iov_list) {
+            delete[] iov_list;
+        }
+        bl.clear();
+        lderr(worker->cct) << __func__ << " fd: " << conn_fd << " send failure: " << UCS_PTR_STATUS(req) << dendl;
+        return UCS_PTR_STATUS(req);
+    } else {
+        req->conn = this;
+        req->bl->swap(bl);
+        req->iov_list = iov_list;
+        req->sn = msg.seq;
+        recv_reqs.insert(req);
+        ldout(worker->cct, 20) << " recv sn: " << msg.seq << ", pending sn: " << sn_recv << dendl;
+    }
+    return UCS_INPROGRESS;
+}
+
+ssize_t UCXConnectedSocketImpl::read_buffers(char *rbuf, size_t bytes)
+{
+    int fd = conn_fd;
+    size_t read_size = 0;
+    size_t left;
+    size_t len;
+
+    auto pb = std::cbegin(rx_bl.buffers());
+    while (read_size < bytes) {
+        left = pb->length();
+        len = bytes - read_size;
+        ldout(cct, 20) << __func__ << " fd: " << fd << " read to " << (void *)rbuf << " wanted " << len << " left " <<
+            left << " pb->>offset= " << pb->offset() << " pb->>length= " << pb->length() << dendl;
+
+        len = len > left ? left : len;
+        memcpy(rbuf + read_size, pb->c_str(), len);
+        read_size += len;
+        pb++;
+    }
+    return read_size;
+}
+
+ssize_t UCXConnectedSocketImpl::zero_copy_read(ceph::bufferlist &bl, size_t length)
+{
+    ceph_assert(worker->center.in_thread());
+    if (rx_bl.length() == 0) {
+	while (ucp_worker_progress(ucp_worker));
+    }
+    size_t bytes = length - bl.length();
+    size_t len = bytes > rx_bl.length() ? rx_bl.length() : bytes;
+    if (len == 0) {
+        if (conn_state >= DISCONNECTED) {
+            return 0; // peer close
+        }
+        return -EAGAIN;
+    }
+    ldout(cct, 20) << __func__ << " fd: " << conn_fd << " rx_bl.length() " << rx_bl.length() << " wanted " << bytes <<
+        " len = " << len << dendl;
+    ceph::bufferlist rbl;
+    rx_bl.splice(0, len, &rbl);
+    bl.append(std::move(rbl));
+    if (rx_bl.length()) {
+        manager->notify(conn_fd, EVENT_READABLE);
+    }
+
+    return len;
+}
+
+ssize_t UCXConnectedSocketImpl::read(char *rbuf, size_t bytes)
+{
+    ceph_assert(worker->center.in_thread());
+    if (rx_bl.length() == 0) {
+	while (ucp_worker_progress(ucp_worker));
+    }
+    size_t len = bytes > rx_bl.length() ? rx_bl.length() : bytes;
+    if (len == 0) {
+        if (conn_state >= DISCONNECTED) {
+            return 0; // peer close
+        }
+        return -EAGAIN;
+    }
+
+    read_buffers(rbuf, len);
+    rx_bl.splice(0, len);
+
+    if (rx_bl.length()) {
+        manager->notify(conn_fd, EVENT_READABLE);
+    }
+
+    return len;
+}
+
+bool UCXConnectedSocketImpl::merge_out_of_order()
+{
+    bool merged = false;
+    if (out_of_order_map.empty()) {
+        return merged;
+    }
+
+    for (auto it = out_of_order_map.begin(); it != out_of_order_map.end();) {
+        auto &bl = it->second;
+        auto seg_beg = it->first;
+        auto seg_len = bl.length();
+        auto seg_end = seg_beg + seg_len;
+
+        if (seg_beg <= sn_recv && seg_end > sn_recv) {
+            // This segment has been received out of order and its previous
+            // segment has been received now
+            auto trim = sn_recv - seg_beg;
+            if (trim) {
+                bl.splice(0, trim);
+                seg_len -= trim;
+            }
+            sn_recv += seg_len;
+            rx_bl.append(std::move(bl));
+            ldout(cct, 20) << __func__ << " merge out of oder sn " << seg_beg << " length" << seg_len <<
+                " current sn: " << sn_recv << " trim " << trim << dendl;
+            // Since c++11, erase() always returns the value of the following element
+            it = out_of_order_map.erase(it);
+            merged = true;
+        } else if (sn_recv >= seg_end) {
+            ldout(cct, 20) << __func__ << " drop merge out of oder seg_beg " << seg_beg << " seg_len: " << seg_len <<
+                " current sn: " << sn_recv << dendl;
+
+            // This segment has been receive already, drop it
+            it = out_of_order_map.erase(it);
+        } else {
+            // seg_beg > sn_recv, can not merge. Note, seg_beg can grow only,
+            // so we can stop looking here.
+            it++;
+            break;
+        }
+    }
+    return merged;
+}
+
+#undef dout_prefix
+#define dout_prefix *_dout << "UCXServerSocketImpl "
+
+UCXServerSocketImpl::UCXServerSocketImpl(UCXWorker *w, int server_socket, entity_addr_t &a, unsigned addr_slot)
+    : ServerSocketImpl(a.get_type(), addr_slot), worker(w), server_fd(server_socket)
+{}
+
+UCXServerSocketImpl::~UCXServerSocketImpl()
+{
+    if (ucp_listener) {
+        ucp_listener_destroy(ucp_listener);
+        ucp_listener = nullptr;
+    }
+    worker->erase_listen_fd(server_fd);
+}
+
+int UCXServerSocketImpl::accept(ConnectedSocket *sock, const SocketOptions &opt, entity_addr_t *peer_addr, Worker *w)
+{
+    ceph_assert(worker->center.in_thread());
+    ucp_conn_request_h conn_req = pop_conn_request();
+    if (conn_req == nullptr) {
+        return -EAGAIN;
+    }
+    ldout(worker->cct, 20) << __func__ << " server_fd: " << server_fd << " conn_req: " << (void *)conn_req << dendl;
+
+    ucp_conn_request_attr_t attr;
+    attr.field_mask = UCP_CONN_REQUEST_ATTR_FIELD_CLIENT_ADDR;
+    ucs_status_t status = ucp_conn_request_query(conn_req, &attr);
+    if (status != UCS_OK) {
+        if (status != UCS_ERR_UNSUPPORTED) {
+            ucp_listener_reject(ucp_listener, conn_req);
+        }
+        return -EIO;
+    };
+    peer_addr->set_sockaddr((sockaddr *)(&attr.client_address));
+    peer_addr->set_type(addr_type);
+    if (!conn_request_queue.empty()) {
+        worker->get_manager()->notify(server_fd, EVENT_READABLE);
+    }
+
+    UCXWorker *dest_worker = static_cast<UCXWorker *>(w);
+    int fd = dest_worker->get_manager()->get_eventfd();
+    UCXConnectedSocketImpl *p = new UCXConnectedSocketImpl(dest_worker, fd);
+    p->set_sockaddr((sockaddr *)(&attr.client_address));
+    if (w->center.in_thread()) {
+        dest_worker->accept_connection(p, conn_req);
+    } else {
+        w->center.submit_to(
+            w->center.get_id(), [dest_worker, p, conn_req]() { dest_worker->accept_connection(p, conn_req); }, true);
+    }
+
+    std::unique_ptr<UCXConnectedSocketImpl> csi(p);
+    *sock = ConnectedSocket(std::move(csi));
+
+    return 0;
+}
+
+void UCXServerSocketImpl::abort_accept()
+{
+    ceph_assert(worker->center.in_thread());
+    if (server_fd == -1)
+        return;
+    if (ucp_listener) {
+        for (auto it = conn_request_queue.cbegin(); it != conn_request_queue.cend(); ++it) {
+            ucp_listener_reject(ucp_listener, *it);
+        }
+        conn_request_queue.clear();
+
+        worker->get_manager()->close(server_fd);
+
+        // ucp_listener_destroy(ucp_listener);
+
+        // ucp_listener = nullptr;
+    }
+    server_fd = -1;
+}
+
+#undef dout_prefix
+#define dout_prefix *_dout << "UCXWorker "
+
+class C_handle_event : public EventCallback {
+    UCXWorker *worker;
+
+public:
+    explicit C_handle_event(UCXWorker *w) : worker(w) {}
+    void do_request(uint64_t fd_or_id) override
+    {
+        worker->handle_poll(fd_or_id);
+    }
+};
+
+void UCXWorker::handle_poll(uint64_t fd_or_id) {}
+
+UCXWorker::UCXWorker(CephContext *c, unsigned i) : Worker(c, i)
+{
+    event_handler = new C_handle_event(this);
+}
+
+UCXWorker::~UCXWorker()
+{
+    for (auto it : conn_map) {
+        UCXConnectedSocketImpl *conn = it.second;
+        conn->close();
+    }
+    conn_map.clear();
+}
+
+void UCXWorker::ep_error_cb_client(void *arg, ucp_ep_h ep, ucs_status_t status)
+{
+    UCXWorker *worker = reinterpret_cast<UCXWorker *>(arg);
+    UCXConnectedSocketImpl *conn = worker->find_connection(ep);
+    if (!conn) {
+        ldout(worker->cct, 0) << "Can not find ep in conn_map" << dendl;
+        return;
+    }
+    CephContext *cct = conn->get_cct();
+    if (conn->is_connected()) {
+        lderr(cct) << "ep_error_cb_client " << ucs_status_string(status) << " fd " << conn->fd() << " ep " << ep <<
+            " connected " << dendl;
+        ceph_assert(UCS_ERR_IO_ERROR != status);
+        conn->set_connect_stat(UCXConnectedSocketImpl::DISCONNECTED);
+        worker->manager->notify(conn->fd(), EVENT_READABLE);
+    } else {
+        lderr(cct) << "ep_error_cb_client " << ucs_status_string(status) << " fd " << conn->fd() << " ep " << ep <<
+            " unconnected" << dendl;
+        ucs_status_ptr_t request = ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FORCE);
+        if (nullptr == request) {
+            ldout(cct, 10) << __func__ << " ucp ep fd: " << conn->fd() << " closed in place..." << dendl;
+        } else if (UCS_PTR_IS_PTR(request)) {
+            /*
+             * We don't care even the request is still in
+             * UCS_INPROGRESS state - it's UCX responsility now
+             */
+            ucp_request_free(request);
+        } else if (UCS_PTR_STATUS(request) != UCS_OK) {
+            lderr(cct) << __func__ << " ep: " << ep << " ucp_ep_close_nb call failed: err " <<
+                ucs_status_string(UCS_PTR_STATUS(request)) << dendl;
+        }
+        ceph_assert(conn->send_reqs.size() == 0);
+        ceph_assert(conn->recv_reqs.size() == 0);
+    }
+}
+
+void UCXWorker::ep_error_cb(void *arg, ucp_ep_h ep, ucs_status_t status)
+{
+    UCXWorker *worker = reinterpret_cast<UCXWorker *>(arg);
+    UCXConnectedSocketImpl *conn = worker->find_connection(ep);
+    if (!conn) {
+        ldout(worker->cct, 0) << "Can not find ep in conn_map" << dendl;
+        return;
+    }
+    CephContext *cct = conn->get_cct();
+    if (conn->is_connected()) {
+        lderr(cct) << "ep_error_cb " << ucs_status_string(status) << " fd " << conn->fd() << " ep " << ep <<
+            " connected " << dendl;
+        ceph_assert(UCS_ERR_IO_ERROR != status);
+        conn->set_connect_stat(UCXConnectedSocketImpl::DISCONNECTED);
+        worker->manager->notify(conn->fd(), EVENT_READABLE);
+    } else {
+        lderr(cct) << "ep_error_cb " << ucs_status_string(status) << " fd " << conn->fd() << " ep " << ep <<
+            " unconnected" << dendl;
+        ucs_status_ptr_t request = ucp_ep_close_nb(ep, UCP_EP_CLOSE_MODE_FORCE);
+        if (nullptr == request) {
+            ldout(cct, 10) << __func__ << " ucp ep fd: " << conn->fd() << " closed in place..." << dendl;
+        } else if (UCS_PTR_IS_PTR(request)) {
+            /*
+             * We don't care even the request is still in
+             * UCS_INPROGRESS state - it's UCX responsility now
+             */
+            ucp_request_free(request);
+        } else if (UCS_PTR_STATUS(request) != UCS_OK) {
+            lderr(cct) << __func__ << " ep: " << ep << " ucp_ep_close_nb call failed: err " <<
+                ucs_status_string(UCS_PTR_STATUS(request)) << dendl;
+        }
+        ceph_assert(conn->send_reqs.size() == 0);
+        ceph_assert(conn->recv_reqs.size() == 0);
+    }
+}
+
+void UCXWorker::accept_connection(UCXConnectedSocketImpl *conn, ucp_conn_request_h conn_req)
+{
+    ceph_assert(center.in_thread());
+    int fd = conn->fd();
+    ucp_ep_h ucp_ep;
+    ucp_ep_params_t ep_params;
+
+    ceph_assert(fd > 0);
+
+    ldout(cct, 20) << __func__ << " fd: " << fd << " ep addr: " << conn_req << dendl;
+
+    ep_params.conn_request = conn_req;
+
+    ep_params.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLER | UCP_EP_PARAM_FIELD_CONN_REQUEST;
+
+    // ep_params.err_mode = UCP_ERR_HANDLING_MODE_PEER;| UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE
+    ep_params.err_handler.cb = ep_error_cb;
+    ep_params.err_handler.arg = reinterpret_cast<void *>(this);
+
+    ucs_status_t status = ucp_ep_create(ucp_worker, &ep_params, &ucp_ep);
+    if (status == UCS_OK) {
+        ldout(cct, 1) << __func__ << " fd: " << fd << " UCP ep: " << (void *)ucp_ep << " successfully created" <<
+            dendl;
+        conn->ucp_ep = ucp_ep;
+        conn->set_connect_stat(UCXConnectedSocketImpl::CONNECTED);
+        conn_map[ucp_ep] = conn;
+        manager->notify(fd, EVENT_READABLE | EVENT_WRITABLE);
+    } else {
+        lderr(cct) << __func__ << " failed to create UCP endpoint fd: " << fd << " " << ucs_status_string(status) <<
+            dendl;
+        conn->set_connect_stat(UCXConnectedSocketImpl::DISCONNECTED);
+        manager->notify(fd, EVENT_READABLE);
+    }
+}
+
+ucs_status_t UCXWorker::am_recv_callback(void *arg, const void *header, size_t header_length, void *data, size_t length,
+    const ucp_am_recv_param_t *param)
+{
+    UCXWorker *worker = reinterpret_cast<UCXWorker *>(arg);
+    ceph_assert(worker->center.in_thread());
+    if (!(param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP)) {
+        ldout(worker->cct, 0) << "UCP_AM_RECV_ATTR_FIELD_REPLY_EP not set" << dendl;
+        return UCS_OK;
+    }
+
+    UCXConnectedSocketImpl *conn = worker->find_connection(param->reply_ep);
+    if (!conn) {
+        ldout(worker->cct, 0) << "Can not find ep in conn_map" << dendl;
+        return UCS_OK;
+    }
+    if (!(param->recv_attr & (UCP_AM_RECV_ATTR_FLAG_DATA | UCP_AM_RECV_ATTR_FLAG_RNDV))) {
+        ldout(worker->cct, 0) <<
+            "Neither UCP_AM_RECV_ATTR_FIELD_DATA nor UCP_AM_RECV_ATTR_FLAG_RNDV is set, just wakeup fd " <<
+            conn->fd() << dendl;
+        return UCS_OK;
+    }
+
+    if (!conn->is_connected()) {
+        ldout(worker->cct, 0) << "Received am from after connection closed " << conn->get_connect_stat() << " fd " <<
+            conn->fd() << dendl;
+        return UCS_OK;
+    }
+    MsgHeader msg = *reinterpret_cast<const MsgHeader *>(header);
+    msg.ntoh();
+    ldout(worker->cct, 20) << "got io (AM) message "
+                           << "ver " << msg.ver << ", "
+                           << "sn " << msg.seq << ", "
+                           << "data len " << msg.data_len << ", "
+                           << " length " << length << dendl;
+
+    // ceph_assert(msg.data_len == length);
+    if (param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV) {
+        return conn->process_recv_request(worker->ucp_worker, data, length, msg);
+    }
+
+    if (param->recv_attr & UCP_AM_RECV_ATTR_FLAG_DATA) {
+        bufferptr bptr(buffer::claim_buffer(length, (char *)data,
+            make_deleter([worker, conn, data, length] { ucp_am_data_release(worker->ucp_worker, data); })));
+        bufferlist bl;
+        bl.append(std::move(bptr));
+        // bufferlist bl = conn->create_bufferlist_with_hint(length, msg, true);
+
+        // bl.append((char *)data, length);
+        conn->insert_recv_buffers(msg.seq, bl, msg.data_len);
+        ldout(worker->cct, 20) << " recv sn: " << msg.seq << ", current sn: " << conn->sn_recv << dendl;
+        return UCS_INPROGRESS;
+    }
+    return UCS_OK;
+}
+
+void UCXWorker::accept_cb(ucp_conn_request_h conn_req, void *arg)
+{
+    UCXServerSocketImpl *server_socket = reinterpret_cast<UCXServerSocketImpl *>(arg);
+    int server_fd = server_socket->fd();
+    UCXWorker *worker = server_socket->get_worker();
+    UCXServerSocketImpl *listern = worker->find_listener(server_fd);
+
+    ceph_assert(listern == server_socket);
+    server_socket->add_conn_request(conn_req);
+    worker->manager->notify(server_fd, EVENT_READABLE);
+
+    ldout(worker->cct, 20) << __func__ << " server_fd: " << server_fd << " UCP conn_req: " << (void *)conn_req << dendl;
+}
+
+int UCXWorker::listen(entity_addr_t &sa, unsigned addr_slot, const SocketOptions &opt, ServerSocket *sock)
+{
+    int server_fd;
+    ucp_listener_params_t params;
+    ceph_assert(center.in_thread());
+    server_fd = manager->get_eventfd();
+    UCXServerSocketImpl *p = new UCXServerSocketImpl(this, server_fd, sa, addr_slot);
+
+    params.field_mask = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR | UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+
+    params.sockaddr.addr = sa.get_sockaddr();
+    params.sockaddr.addrlen = sizeof(struct sockaddr);
+
+    params.conn_handler.cb = accept_cb;
+    params.conn_handler.arg = reinterpret_cast<void *>(p);
+
+    ucp_listener_h ucp_listener;
+
+    /* Create a listener on the server side to listen on the given address. */
+    ucs_status_t status = ucp_listener_create(ucp_worker, &params, &ucp_listener);
+    if (UCS_OK != status) {
+        delete p;
+        manager->close(server_fd);
+        lderr(cct) << __func__ << " Failed to create UCP listener with status: " << ucs_status_string(status) << dendl;
+        return -EBUSY;
+    }
+
+    ucp_listener_attr_t listen_attr;
+    listen_attr.field_mask = UCP_LISTENER_ATTR_FIELD_SOCKADDR;
+
+    status = ucp_listener_query(ucp_listener, &listen_attr);
+    if (status != UCS_OK) {
+        delete p;
+        manager->close(server_fd);
+        lderr(cct) << __func__ << "<Failed to query UCP listener> to addr: " << sa.get_sockaddr() << " status: " <<
+            ucs_status_string(status) << dendl;
+        ucp_listener_destroy(ucp_listener);
+        return -EIO;
+    }
+
+    ldout(cct, 0) << __func__ << " Created listner to addr: " << sa.get_sockaddr() << " server_fd: " << server_fd <<
+        dendl;
+
+    p->set_ucp_listener(ucp_listener);
+    listen_map[server_fd] = p;
+    *sock = ServerSocket(std::unique_ptr<ServerSocketImpl>(p));
+    return 0;
+}
+
+int UCXWorker::connect(const entity_addr_t &peer_addr, const SocketOptions &opts, ConnectedSocket *sock)
+{
+    ceph_assert(center.in_thread());
+    int fd = manager->get_eventfd();
+
+    ucp_ep_params_t ep_params;
+    ep_params.flags = UCP_EP_PARAMS_FLAGS_CLIENT_SERVER;
+    ep_params.field_mask = UCP_EP_PARAM_FIELD_FLAGS | UCP_EP_PARAM_FIELD_SOCK_ADDR | UCP_EP_PARAM_FIELD_ERR_HANDLER |
+        UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
+    ep_params.err_mode = UCP_ERR_HANDLING_MODE_PEER;
+    ep_params.err_handler.cb = ep_error_cb_client;
+    ep_params.err_handler.arg = reinterpret_cast<void *>(this);
+
+    ep_params.sockaddr.addr = peer_addr.get_sockaddr();
+    ep_params.sockaddr.addrlen = sizeof(struct sockaddr);
+    entity_addr_t addr = opts.connect_bind_addr;
+    if (cct->_conf->ms_bind_before_connect && (!addr.is_blank_ip())) {
+        addr.set_port(0);
+        ep_params.field_mask |= UCP_EP_PARAM_FIELD_LOCAL_SOCK_ADDR;
+        ep_params.local_sockaddr.addr = addr.get_sockaddr();
+        ep_params.local_sockaddr.addrlen = addr.get_sockaddr_len();
+    }
+    ucp_ep_h ucp_ep;
+    ucs_status_t status = ucp_ep_create(ucp_worker, &ep_params, &ucp_ep);
+    if (UCS_OK != status) {
+        manager->close(fd);
+        lderr(cct) << __func__ << " creating UCP ep to addr failed addr: " << peer_addr.get_sockaddr() << " status: " <<
+            ucs_status_string(status) << dendl;
+        return -EIO;
+    }
+    UCXConnectedSocketImpl *conn = new UCXConnectedSocketImpl(this, fd);
+    conn->ucp_ep = ucp_ep;
+    conn->set_connect_stat(UCXConnectedSocketImpl::CONNECTED);
+    conn_map[ucp_ep] = conn;
+    ldout(cct, 10) << __func__ << " fd: " << conn->fd() << " UCP ep: " << (void *)ucp_ep << " to addr: " <<
+        peer_addr.get_sockaddr() << " successfully created" << dendl;
+
+    std::unique_ptr<UCXConnectedSocketImpl> csi(conn);
+    *sock = ConnectedSocket(std::move(csi));
+
+    return 0;
+}
+
+void UCXWorker::set_recv_handler(void)
+{
+    ucp_am_handler_param_t param;
+    param.field_mask = UCP_AM_HANDLER_PARAM_FIELD_ID | UCP_AM_HANDLER_PARAM_FIELD_CB | UCP_AM_HANDLER_PARAM_FIELD_ARG |
+        UCP_AM_HANDLER_PARAM_FIELD_FLAGS;
+    param.id = UCXConnectedSocketImpl::AM_ID;
+    param.cb = UCXWorker::am_recv_callback;
+    param.arg = this;
+    param.flags = UCP_AM_FLAG_PERSISTENT_DATA;
+    ceph_assert(UCS_OK == ucp_worker_set_am_recv_handler(ucp_worker, &param));
+}
+
+void UCXWorker::initialize()
+{
+    ucs_status_t status;
+    ucp_worker_params_t params;
+    params.thread_mode = UCS_THREAD_MODE_MULTI;
+    params.field_mask = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    UCXEventDriver *driver = dynamic_cast<UCXEventDriver *>(center.get_driver());
+    manager = driver->get_manager();
+    status = ucp_worker_create(ucp_context, &params, &ucp_worker);
+    if (UCS_OK != status) {
+        lderr(cct) << __func__ << " failed to create UCP worker " << dendl;
+        ceph_abort();
+    }
+
+    status = ucp_worker_get_efd(ucp_worker, &ucp_fd);
+    if (UCS_OK != status) {
+        lderr(cct) << __func__ << " failed to obtain UCP worker event fd " << dendl;
+        ceph_abort();
+    }
+
+    driver->driver_init(ucp_worker, ucp_fd);
+
+    set_recv_handler();
+    center.create_file_event(ucp_fd, EVENT_READABLE, event_handler);
+}
+
+void UCXWorker::destroy()
+{
+    if (ucp_fd > 0) {
+        ldout(cct, 20) << __func__ << dendl;
+
+        ceph_assert(nullptr != ucp_worker);
+
+        center.delete_file_event(ucp_fd, EVENT_READABLE);
+        delete event_handler;
+        event_handler = nullptr;
+        ucp_worker_destroy(ucp_worker);
+        ucp_worker = nullptr;
+        ucp_fd = -1;
+    }
+}
+
+#undef dout_prefix
+#define dout_prefix *_dout << "UCXStack "
+
+void UCXStack::ucx_contex_create()
+{
+    ucp_params_t params;
+    ucs_status_t status;
+
+    ucp_config_t *ucp_config;
+
+    if (nullptr != ucp_context) {
+        return;
+    }
+
+    ldout(cct, 10) << __func__ << " UCX contex is going to be created..." << dendl;
+
+    int rc = setenv("UCX_NET_DEVICES", cct->_conf->ms_async_ucx_device.c_str(), 1);
+    if (rc) {
+        lderr(cct) << __func__ << " failed to export UCX_CEPH_NET_DEVICES. Application aborts." << dendl;
+        ceph_abort();
+    }
+
+    setenv("UCX_MEM_MMAP_HOOK_MODE", "none", 1);
+    setenv("UCX_MAX_EAGER_RAILS", "5", 1);
+    setenv("UCX_MAX_RNDV_RAILS", "5", 1);
+    setenv("UCX_PROTO_ENABLE", "y", 1);
+    setenv("UCX_WARN_UNUSED_ENV_VARS", "n", 1);
+    rc = setenv("UCX_LOG_LEVEL", cct->_conf->ms_async_ucx_log_level.c_str(), 1);
+    if (rc) {
+        lderr(cct) << __func__ << " failed to export UCX_LOG_LEVEL. Application aborts." << dendl;
+        ceph_abort();
+    }
+
+    rc = setenv("UCX_RNDV_THRESH", cct->_conf->ms_async_ucx_rndv_thresh.c_str(), 1);
+    if (rc) {
+        lderr(cct) << __func__ << " failed to export UCX_CEPH_RNDV_THRESH. Application aborts." << dendl;
+        ceph_abort();
+    }
+
+    rc = setenv("UCX_TLS", cct->_conf->ms_async_ucx_tls.c_str(), 1);
+    if (rc) {
+        lderr(cct) << __func__ << " failed to export UCX_CEPH_TLS. Application aborts." << dendl;
+        ceph_abort();
+    }
+
+    std::string addr = cct->_conf.get_val<std::string>("rdma_cm_source_addr");
+    if (!addr.empty()) {
+        rc = setenv("UCX_RDMA_CM_SOURCE_ADDRESS", addr.c_str(), 1);
+        if (rc) {
+            lderr(cct) << __func__ << " failed to export UCX_RDMA_CM_SOURCE_ADDRESS. Application aborts." << dendl;
+            ceph_abort();
+        }
+        lderr(cct) << __func__ << " export UCX_RDMA_CM_SOURCE_ADDRESS = " << addr << dendl;
+    }
+
+    status = ucp_config_read("", nullptr, &ucp_config);
+    if (UCS_OK != status) {
+        lderr(cct) << __func__ << "failed to read UCP config" << dendl;
+        ceph_abort();
+    }
+
+    memset(&params, 0, sizeof(params));
+    params.field_mask = UCP_PARAM_FIELD_FEATURES | UCP_PARAM_FIELD_REQUEST_SIZE | UCP_PARAM_FIELD_REQUEST_INIT |
+        UCP_PARAM_FIELD_REQUEST_CLEANUP | UCP_PARAM_FIELD_TAG_SENDER_MASK | UCP_PARAM_FIELD_MT_WORKERS_SHARED;
+
+    params.features = UCP_FEATURE_WAKEUP | UCP_FEATURE_AM;
+
+    params.mt_workers_shared = 1;
+    params.tag_sender_mask = -1;
+    params.request_size = sizeof(UCXReqDescr);
+
+    params.request_init = UCXConnectedSocketImpl::request_init;
+    params.request_cleanup = UCXConnectedSocketImpl::request_cleanup;
+
+    status = ucp_init(&params, ucp_config, &ucp_context);
+    ucp_config_release(ucp_config);
+
+    if (UCS_OK != status) {
+        lderr(cct) << __func__ << "failed to init UCP context" << dendl;
+        ceph_abort();
+    }
+
+    ucp_context_print_info(ucp_context, stdout);
+}
+
+UCXStack::UCXStack(CephContext *cct) : NetworkStack(cct)
+{
+    ldout(cct, 10) << __func__ << " constructing UCX stack "
+                   << " with " << get_num_worker() << " workers " << dendl;
+}
+
+UCXStack::~UCXStack()
+{
+    if (nullptr != ucp_context) {
+        ucp_cleanup(ucp_context);
+    }
+}
+
+void UCXStack::spawn_worker(Worker* worker, std::function<void()> &&func)
+{
+    ucx_contex_create();
+    UCXWorker *w = (UCXWorker *)worker;
+    w->set_ucp_context(ucp_context);
+    threads.emplace_back(std::move(func));
+}
+
+void UCXStack::join_worker(unsigned i)
+{
+    ceph_assert(threads.size() > i && threads[i].joinable());
+    threads[i].join();
+}
+
diff --git a/src/msg/async/ucx/UCXStack.h b/src/msg/async/ucx/UCXStack.h
new file mode 100644
index 000000000..5f4f3fcef
--- /dev/null
+++ b/src/msg/async/ucx/UCXStack.h
@@ -0,0 +1,336 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (c) Huawei Technologies Co., Ltd. 2023-2023.
+ *
+ * Author: Chunsong Feng <fengchunsong@huawei.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef CEPH_UCXSTACK_H
+#define CEPH_UCXSTACK_H
+
+#include <thread>
+#include <deque>
+
+#include "common/ceph_context.h"
+#include "common/debug.h"
+#include "common/errno.h"
+
+#include "msg/async/Stack.h"
+#include "UCXEvent.h"
+
+extern "C" {
+#include <ucp/api/ucp.h>
+};
+
+class UCXStack;
+class UCXConnectedSocketImpl;
+class UCXServerSocketImpl;
+
+struct MsgHeader{
+    uint16_t ver = 1;
+    uint16_t tag = 0;
+    uint32_t data_len;
+    uint64_t seq;
+    uint32_t align = 0xFFFFFFFF;  
+    void hton(void) {
+        ver = htole16(ver);
+        tag = htole16(tag);
+        seq = htole64(seq);        
+        data_len = htole64(data_len);        
+        align = htole32(align);       
+    }
+    void ntoh(void) {
+        ver = le16toh(ver);
+        tag = le16toh(tag);
+        seq = le64toh(seq);
+        data_len = le64toh(data_len);
+        align = le32toh(align);      
+    }
+} __attribute__((packed));
+
+struct UCXReqDescr {
+    UCXConnectedSocketImpl *conn;
+    bufferlist *bl;
+    ucp_dt_iov_t *iov_list;
+    uint64_t sn;
+};
+
+struct desc_data {
+    void *data;
+    size_t length;
+    struct MsgHeader msg;
+};
+
+class UCXWorker : public Worker {
+public:
+    explicit UCXWorker(CephContext *c, unsigned i);
+    virtual ~UCXWorker();
+
+    int listen(entity_addr_t &addr, unsigned addr_slot, const SocketOptions &opts, ServerSocket *) override;
+    int connect(const entity_addr_t &addr, const SocketOptions &opts, ConnectedSocket *socket) override;
+    void initialize() override;
+    void destroy() override;
+
+    void accept_connection(UCXConnectedSocketImpl *conn, ucp_conn_request_h conn_req);
+    static void accept_cb(ucp_conn_request_h conn_req, void *arg);
+    static void ep_error_cb(void *arg, ucp_ep_h ep, ucs_status_t status);
+    static void ep_error_cb_client(void *arg, ucp_ep_h ep, ucs_status_t status);
+
+    static ucs_status_t am_recv_callback(void *arg, const void *header, size_t header_length, void *data, size_t length,
+        const ucp_am_recv_param_t *param);
+
+    ucp_worker_h get_ucp_worker()
+    {
+        return ucp_worker;
+    }
+
+    void set_ucp_context(ucp_context_h ctx)
+    {
+	ucp_context = ctx;
+    }
+
+    UCXEventManager *get_manager()
+    {
+        return manager;
+    }
+
+    void erase_connection(ucp_ep_h ucp_ep)
+    {
+        conn_map.erase(ucp_ep);
+    }
+
+    UCXConnectedSocketImpl *find_connection(ucp_ep_h ucp_ep)
+    {
+        if (conn_map.count(ucp_ep) == 0)
+            return nullptr;
+        return conn_map[ucp_ep];
+    }
+    void erase_listen_fd(int fd)
+    {
+        listen_map.erase(fd);
+    }
+
+    UCXServerSocketImpl *find_listener(int fd)
+    {
+        if (listen_map.count(fd) == 0)
+            return nullptr;
+        return listen_map[fd];
+    }    
+
+    void handle_poll(uint64_t fd_or_id);
+
+private:
+    UCXEventManager *manager;
+    ucp_worker_h ucp_worker = nullptr;
+    ucp_context_h ucp_context = nullptr;
+    EventCallbackRef event_handler;
+
+    std::set<int> accepting;
+    std::unordered_map<ucp_ep_h, UCXConnectedSocketImpl *> conn_map;
+    std::unordered_map<int, UCXServerSocketImpl *> listen_map;
+    int ucp_fd = -1;
+
+    void set_recv_handler(void);
+};
+
+class UCXConnectedSocketImpl : public ConnectedSocketImpl {
+    static constexpr __u16 PAGE_SIZE_ALIGNMENT = 4096;
+    
+    UCXEventManager *manager;
+    bool use_zero_copy;
+public:
+    static constexpr unsigned AM_ID = 0x43657068; // Ceph
+    enum CONNECT_STATE {
+        CONNECTING = 0,
+        CONNECTED,
+        DISCONNECTED,
+        CLOSED,
+    };
+    ucp_ep_h ucp_ep = nullptr;
+    uint64_t sn_send = 0;
+    uint64_t sn_recv = 0;
+    std::set<UCXReqDescr *> send_reqs;
+    std::set<UCXReqDescr *> recv_reqs;
+    CephContext *get_cct()
+    {
+        return cct;
+    }
+    UCXWorker *get_worker()
+    {
+        return worker;
+    }
+
+    UCXConnectedSocketImpl(UCXWorker *w, int fd);
+    virtual ~UCXConnectedSocketImpl();
+
+    int is_connected() override
+    {
+        return (conn_state == CONNECTED) && (ucp_ep != nullptr);
+    };
+    ssize_t read(char *, size_t) override;
+    ssize_t zero_copy_read(ceph::bufferlist &bl, size_t length) override;
+    ssize_t send(bufferlist &bl, bool more) override;
+    void shutdown() override;
+    void close() override;
+    int fd() const override
+    {
+        return conn_fd;
+    }
+    int getsockname(int sockfd, struct sockaddr *addr, socklen_t *addrlen) override;                       
+    //int socket_fd() const override
+    //{
+    //    return conn_fd;
+    //}
+
+    
+    void set_sockaddr(const struct sockaddr *sa){
+        memcpy(&sock_addr, sa, sizeof(sock_addr));
+    }
+    static void request_init(void *req);
+    static void request_cleanup(void *req);
+
+    ssize_t do_send(bufferlist &bl);
+    void conn_release_recvs();
+    void ucx_ep_close(bool close_event);
+
+    static void send_completion_cb(void *req, ucs_status_t status, void *userdata);
+
+    static void am_recv_data_callback(void *req, ucs_status_t status, size_t length, void *user_data);
+    ucs_status_t process_recv_request(ucp_worker_h ucp_worker, void *data, size_t length, MsgHeader &msg);
+
+    ssize_t read_buffers(char *rbuf, size_t bytes);
+
+    bool merge_out_of_order();
+    void insert_recv_buffers(uint64_t sn, bufferlist &bl, size_t length);
+    void insert_out_of_order(uint64_t seg, bufferlist &bl)
+    {
+        auto it = out_of_order_map.find(seg);
+        if (it == out_of_order_map.end())
+        {
+            out_of_order_map.emplace(seg, std::move(bl));
+        }
+        else
+        {
+            it->second.append(std::move(bl));
+        }        
+    }
+    void set_connect_stat(enum CONNECT_STATE stat)
+    {
+        conn_state = stat;
+    }
+    enum CONNECT_STATE get_connect_stat(void)
+    {
+        return conn_state;
+    }
+    bufferlist create_bufferlist_with_hint(size_t length, MsgHeader &msg);
+
+private:
+    CephContext *cct;
+    UCXWorker *worker;
+    ucp_worker_h ucp_worker;
+    bufferlist rx_bl;
+    std::map<uint64_t, bufferlist> out_of_order_map;
+    int conn_fd;
+    enum CONNECT_STATE conn_state = CONNECTING;
+    struct sockaddr sock_addr;
+    int submit(bufferlist &bl, MsgHeader *msg, const void *buffer, size_t count, ucp_datatype_t type);  
+};
+
+class UCXServerSocketImpl : public ServerSocketImpl {
+public:
+    UCXServerSocketImpl(UCXWorker *w, int server_socket, entity_addr_t &a, unsigned addr_slot);
+    ~UCXServerSocketImpl();
+
+    int accept(ConnectedSocket *sock, const SocketOptions &opt, entity_addr_t *out, Worker *w) override;
+    void abort_accept() override;
+    int fd() const override
+    {
+        return server_fd;
+    }
+
+    UCXWorker *get_worker()
+    {
+        return worker;
+    }
+
+    CephContext *cct()
+    {
+        return worker->cct;
+    }
+
+    void add_conn_request(ucp_conn_request_h conn_req)
+    {
+        conn_request_queue.push_back(conn_req);
+    }
+
+    void set_ucp_listener(ucp_listener_h listener)
+    {
+        ucp_listener = listener;
+    }
+private:
+    UCXWorker *worker;
+    int server_fd;
+    std::deque<ucp_conn_request_h> conn_request_queue;
+    ucp_listener_h ucp_listener = nullptr;
+
+    ucp_conn_request_h pop_conn_request()
+    {
+        if (conn_request_queue.empty()) {
+            return nullptr;
+        }
+
+        ucp_conn_request_h conn_req = conn_request_queue.front();
+        conn_request_queue.pop_front();
+        return conn_req;
+    }
+};
+
+class UCXStack : public NetworkStack {
+public:
+    explicit UCXStack(CephContext *cct);
+    virtual ~UCXStack();
+
+    virtual Worker* create_worker(CephContext *c, unsigned worker_id) override {
+        return new UCXWorker(c, worker_id);
+    }
+
+    bool support_zero_copy_read() const override
+    {
+        return true;
+    }
+
+    bool nonblock_connect_need_writable_event() const override
+    {
+        return false;
+    }
+
+    void spawn_worker(Worker* worker, std::function<void()> &&func) override;
+    void join_worker(unsigned i) override;
+
+    bool is_ready() override
+    {
+        return nullptr != ucp_context;
+    };
+
+    ucp_context_h get_ucp_context()
+    {
+        return ucp_context;
+    }
+
+private:
+    std::vector<std::thread> threads;
+    ucp_context_h ucp_context = nullptr;
+
+    void ucx_contex_create();
+};
+
+#endif
diff --git a/src/test/msgr/test_async_networkstack.cc b/src/test/msgr/test_async_networkstack.cc
index 14923fd7d..08f009f5f 100644
--- a/src/test/msgr/test_async_networkstack.cc
+++ b/src/test/msgr/test_async_networkstack.cc
@@ -71,16 +71,17 @@ class NetworkWorkerTest : public ::testing::TestWithParam<const char*> {
   NetworkWorkerTest() {}
   void SetUp() override {
     cerr << __func__ << " start set up " << GetParam() << std::endl;
-    if (strncmp(GetParam(), "dpdk", 4)) {
+    if (strncmp(GetParam(), "ucx", 3)) {
       g_ceph_context->_conf.set_val("ms_type", "async+posix");
       addr = "127.0.0.1:15000";
       port_addr = "127.0.0.1:15001";
     } else {
-      g_ceph_context->_conf.set_val_or_die("ms_dpdk_debug_allow_loopback", "true");
       g_ceph_context->_conf.set_val_or_die("ms_async_op_threads", "2");
-      string ipv4_addr = g_ceph_context->_conf.get_val<std::string>("ms_dpdk_host_ipv4_addr");
-      addr = ipv4_addr + std::string(":15000");
-      port_addr = ipv4_addr + std::string(":15001");
+      g_ceph_context->_conf.set_val_or_die("ms_type", "async+ucx");
+      g_ceph_context->_conf.set_val_or_die("ms_async_ucx_device", "mlx5_0:1");
+      g_ceph_context->_conf.set_val_or_die("ms_async_ucx_tls", "rc_mlx5");
+      addr = "192.168.37.197:15000";
+      port_addr = "192.168.37.197:15001";
     }
     stack = NetworkStack::create(g_ceph_context, GetParam());
     stack->start();
@@ -1059,6 +1060,7 @@ INSTANTIATE_TEST_SUITE_P(
 #ifdef HAVE_DPDK
     "dpdk",
 #endif
+    "ucx",
     "posix"
   )
 );
diff --git a/src/test/msgr/test_msgr.cc b/src/test/msgr/test_msgr.cc
index 8766ac30e..a20dc8818 100644
--- a/src/test/msgr/test_msgr.cc
+++ b/src/test/msgr/test_msgr.cc
@@ -2338,7 +2338,13 @@ INSTANTIATE_TEST_SUITE_P(
   Messenger,
   MessengerTest,
   ::testing::Values(
-    "async+posix"
+    "async+posix",
+#ifdef HAVE_RDMA
+    "async+rdma",
+#endif
+#ifdef HAVE_UCX
+    "async+ucx"
+#endif
   )
 );
 
